{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 282,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "import Bio.PDB\n",
    "import Bio.SeqIO\n",
    "from Bio import pairwise2\n",
    "from Bio.SubsMat import MatrixInfo as matlist\n",
    "from Bio.SeqUtils import seq3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 283,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "reload(sys)\n",
    "sys.setdefaultencoding('utf-8')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# launch pymol\n",
    "\n",
    "import __main__\n",
    "__main__.pymol_argv = ['pymol','-Qc'] # Pymol: quiet and no GUI\n",
    "from time import sleep\n",
    "import pymol\n",
    "from pymol import cmd\n",
    "\n",
    "pymol.finish_launching()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# paths\n",
    "\n",
    "hla_fasta = '../data/mhc_alleles/hla/hla_prot.fasta'\n",
    "pdb_path = '../data/structures'\n",
    "aligned_pdb_path = '../sandbox/aligned_pdb'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 285,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# tables\n",
    "\n",
    "annotation = pd.read_csv('../sandbox/detailed_mhc', sep='\\t')\n",
    "alleles_vs_pdb = pd.read_csv('../sandbox/alleles_vs_pdb', sep='\\t', index_col=0, low_memory=False)\n",
    "peptides_vs_pdb = pd.read_csv('../sandbox/peptides_vs_pdb', sep='\\t', index_col=0, low_memory=False)\n",
    "ligand_assays = pd.read_csv('../data/epitopes/mhc_ligand_full.csv', skiprows=1, low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# table conversion\n",
    "\n",
    "ligand_assays = ligand_assays[ligand_assays['MHC allele class'] == 'I']\n",
    "affinity = ligand_assays[['Description', 'Allele Name', 'Qualitative Measure', 'Quantitative measurement']]\n",
    "affinity['Length'] = [len(x) for x in affinity.Description]\n",
    "affinity = affinity[(affinity.Length < 12) & (affinity.Length > 7)]\n",
    "affinity = affinity[affinity['Allele Name'].str.contains('HLA-[ABC]\\*\\S+$', regex=True)]\n",
    "affinity['Allele Name'] = [x[4:] for x in affinity['Allele Name']]\n",
    "\n",
    "iclass = annotation[annotation.mhc_class == 'I']\n",
    "iclass = iclass[iclass.a_chain_allele.str.contains('[ABC]\\*', regex=True)]\n",
    "iclass.a_chain_allele = [':'.join(x.split(':')[:2]) for x in iclass.a_chain_allele]\n",
    "iclass = iclass.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 292,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# place mhc to the specific position\n",
    "def localize_pdb(name, amhc, bmhc, pept, plen):\n",
    "    # define peptide CA ids\n",
    "    parser = Bio.PDB.PDBParser(QUIET=True)\n",
    "    ppb = Bio.PDB.CaPPBuilder() \n",
    "    struct = parser.get_structure(id=name, file=pdb_path+'/'+name+'.pdb')[0]\n",
    "    reslist = ppb.build_peptides(struct[pept])[0]\n",
    "    firstCA = reslist[0].get_id()[1]\n",
    "    lastCA = reslist[-1].get_id()[1]\n",
    "    \n",
    "    assert(lastCA-firstCA+1 == plen)\n",
    "    \n",
    "    # translate complex by -startCA\n",
    "    cmd.select('sele', '%s//%s/%i/CA %s//%s/%i/CA' % (name, pept, firstCA, name, pept, lastCA))\n",
    "    start, end = cmd.get_coords('sele', 1)\n",
    "    cmd.translate(list(-start), name)\n",
    "    \n",
    "    # rotate complex around (0, 0, 0), so that endCA.y = 0\n",
    "    start, end = cmd.get_coords('sele', 1)\n",
    "    angle = np.arctan(end[1]/end[0])\n",
    "    if end[0] < 0.0:\n",
    "        angle = np.pi + angle\n",
    "    angle = -angle*180.0/np.pi\n",
    "    cmd.rotate('z', angle, name, origin=[0.0, 0.0, 0.0])\n",
    "    \n",
    "    # superpose peptide line on x axis\n",
    "    start, end = cmd.get_coords('sele', 1)\n",
    "    angle = np.arctan(end[2]/end[0])\n",
    "    if end[0] < 0.0:\n",
    "        angle = np.pi + angle\n",
    "    angle = angle*180.0/np.pi\n",
    "    print(cmd.rotate('y', angle, name, origin=[0.0, 0.0, 0.0]))\n",
    "    \n",
    "    # shift along x axis\n",
    "    start, end = cmd.get_coords('sele', 1)\n",
    "    print(cmd.translate([-(start[0]+end[0])/2.0, 0.0, 0.0], name))\n",
    "    \n",
    "    # superpose com on the xz plane\n",
    "    com = cmd.centerofmass('%s//%s// %s//%s//' % (name, amhc, name, bmhc))\n",
    "    angle = np.arctan(com[2]/com[1])\n",
    "    if com[1] < 0.0:\n",
    "        angle = np.pi + angle\n",
    "    angle = -(angle)*180.0/np.pi\n",
    "    print(cmd.rotate('x', angle, name, origin=[0.0, 0.0, 0.0]))\n",
    "    \n",
    "    # verify the result\n",
    "    com = cmd.centerofmass('%s//%s// %s//%s//' % (name, amhc, name, bmhc))\n",
    "    start, end = cmd.get_coords('sele', 1)\n",
    "    assert(com[2] < 0.001)\n",
    "    assert(start[2] < 0.001 and start[1] < 0.001)\n",
    "    assert(end[2] < 0.001 and end[1] < 0.001)\n",
    "    \n",
    "    #cmd.delete('!(%s//%s// %s//%s// %s//%s//)' % (name, mhca, name, mhcb, name, pept))\n",
    "    \n",
    "    #cmd.save('../sandbox/positioned_pdb/%s_posed.pdb' % name, \n",
    "    #         '%s//%s// %s//%s// %s//%s//' % (name, mhca, name, mhcb, name, pept))\n",
    "    \n",
    "    sleep(0.5)\n",
    "\n",
    "# name should be preloaded to pymol with cmd.load()    \n",
    "def localize_pdb_wrapper(name, table):\n",
    "    line = table[table.pdb == name].iloc[0, :]\n",
    "    mhca = line['a_chain_id']\n",
    "    mhcb = line['b_chain_id']\n",
    "    pept = line['antigen_id']\n",
    "    plen = int(line['antigen_len'])\n",
    "\n",
    "    localize_pdb(name, mhca, mhcb, pept, plen)\n",
    "    \n",
    "# name should be preloaded to pymol with cmd.load()    \n",
    "def clean(name, table):\n",
    "    line = table[table.pdb == name].iloc[0, :]\n",
    "    mhca = line['a_chain_id']\n",
    "    mhcb = line['b_chain_id']\n",
    "    pept = line['antigen_id']\n",
    "    \n",
    "    cmd.delete('!(%s//%s// %s//%s// %s//%s//)' % (name, mhca, name, mhcb, name, pept))\n",
    "    \n",
    "# reference and mobile should be preloaded to pymol with cmd.load()\n",
    "def align_mob2ref(mob_name, ref_name, table):\n",
    "    line = table[table.pdb == ref_name].iloc[0, :]\n",
    "    ref_mhca = line['a_chain_id']\n",
    "    ref_mhcb = line['b_chain_id']\n",
    "    ref_pept = line['antigen_id']\n",
    "    \n",
    "    line = table[table.pdb == mob_name].iloc[0, :]\n",
    "    mob_mhca = line['a_chain_id']\n",
    "    mob_mhcb = line['b_chain_id']\n",
    "    mob_pept = line['antigen_id']\n",
    "\n",
    "    cmd.align('%s//%s// %s//%s//' % (mob_name, mob_mhca, mob_name, mob_pept), \n",
    "              '%s//%s// %s//%s//' % (ref_name, ref_mhca, ref_name, ref_pept))\n",
    "    \n",
    "# name should be preloaded to pymol with cmd.load()\n",
    "def save2pdb(name, table, path):\n",
    "    line = table[table.pdb == name].iloc[0, :]\n",
    "    mhca = line['a_chain_id']\n",
    "    mhcb = line['b_chain_id']\n",
    "    pept = line['antigen_id']\n",
    "    \n",
    "    cmd.save(path, '%s//%s// %s//%s// %s//%s//' % (name, mhca, name, mhcb, name, pept)) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 300,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def align_all_pdbs(table):\n",
    "    ref=table.loc[0, 'pdb']\n",
    "\n",
    "    cmd.reinitialize()\n",
    "    cmd.load(pdb_path+'/'+ref+'.pdb', ref)\n",
    "    clean(ref, table)\n",
    "    localize_pdb_wrapper(ref, table)\n",
    "\n",
    "    for i in range(1, table.shape[0]):\n",
    "        name=table.loc[i, 'pdb']\n",
    "        cmd.load(pdb_path+'/'+name+'.pdb', name)\n",
    "        clean(name, table)\n",
    "        align_mob2ref(name, ref, table)\n",
    "        save2pdb(name, table, '%s/%s_aligned.pdb' % (aligned_pdb_path, name))\n",
    "        cmd.delete(name)\n",
    "\n",
    "    save2pdb(ref, table, '%s/%s_aligned.pdb' % (aligned_pdb_path, ref))\n",
    "    cmd.delete(ref)\n",
    "        \n",
    "#align_all_pdbs(iclass)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 336,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# s1 - mutated sequence\n",
    "# s2 - reference sequence\n",
    "def align(s1, s2):\n",
    "    matrix = matlist.blosum62\n",
    "    gap_open = -100.0\n",
    "    gap_extend = -0.5\n",
    "    alns = pairwise2.align.globalds(s1, s2, matrix, gap_open, gap_extend)\n",
    "    top_aln = alns[0]\n",
    "    \n",
    "    \n",
    "    print(top_aln[0])\n",
    "    print(top_aln[1])\n",
    "    \n",
    "    # check that there is no gaps inside the alignment\n",
    "    assert(re.search('[A-Z]-+[A-Z]', top_aln[0]) == None)\n",
    "    assert(re.search('[A-Z]-+[A-Z]', top_aln[1]) == None)\n",
    "    \n",
    "    mutations = []\n",
    "    gap = re.match('(-+)', top_aln[0])\n",
    "    shift = 0\n",
    "    if gap:\n",
    "        shift = gap.end(1) - gap.start(1)\n",
    "        \n",
    "    for i, a1, a2 in zip(list(range(-shift, len(top_aln[0])-shift)), top_aln[0], top_aln[1]):\n",
    "        if a1 != '-' and a2 != '-':\n",
    "            if a1 != a2:\n",
    "                mutations.append((i, a1, a2))\n",
    "    return mutations\n",
    "\n",
    "# mutate pdb structure in accordance to input sequences\n",
    "# here we assume that all chains are continuous: CHECK THIS\n",
    "def mutate(pdb, newseq, newpep, table):\n",
    "    parser = Bio.PDB.PDBParser(QUIET=True)\n",
    "    ppb=Bio.PDB.PPBuilder()\n",
    "    \n",
    "    name = pdb; \n",
    "    amhc = table[table.pdb == pdb].iloc[0, 3]\n",
    "    pept = table[table.pdb == pdb].iloc[0, 7]\n",
    "    b2m = table[table.pdb == pdb].iloc[0, 5]\n",
    "    \n",
    "    struct = parser.get_structure(id=name, file=pdb_path+'/'+name+'.pdb')[0]\n",
    "    oldseq = ''.join([str(x.get_sequence()) for x in ppb.build_peptides(struct[amhc])])\n",
    "    oldpep = table[table.pdb == pdb].iloc[0, 8]\n",
    "    \n",
    "    assert(len(oldpep) == len(newpep))\n",
    "    mhc_mutations = align(oldseq, newseq)\n",
    "    pep_mutations = align(oldpep, newpep)\n",
    "    \n",
    "    # remove alternative atom locations\n",
    "    cmd.remove(\"not alt ''+A\")\n",
    "    cmd.alter('all', \"alt=''\")\n",
    "    #cmd.save('../sandbox/mutated/original.pdb', '%s//%s// %s//%s// %s//%s//' % (name, amhc, name, pept, name, b2m))\n",
    "    \n",
    "    # mutate\n",
    "    cmd.wizard('mutagenesis')\n",
    "    cmd.refresh_wizard()\n",
    "\n",
    "    first_aa_id = ppb.build_peptides(struct[amhc])[0][0].get_id()[1]\n",
    "    for i, old, new in mhc_mutations:\n",
    "        print(i, old, new)\n",
    "        cmd.get_wizard().do_select(amhc+'/'+str(i+first_aa_id)+'/')\n",
    "        cmd.get_wizard().set_mode(seq3(new).upper())\n",
    "        cmd.get_wizard().apply()\n",
    "    print(\"\\n%s: %i MHC mutations\\n\" % (name, len(mhc_mutations)))\n",
    "        \n",
    "    first_aa_id = ppb.build_peptides(struct[pept])[0][0].get_id()[1]\n",
    "    for i, old, new in pep_mutations:\n",
    "        print(i, old, new)\n",
    "        cmd.get_wizard().do_select(pept+'/'+str(i+first_aa_id)+'/')\n",
    "        cmd.get_wizard().set_mode(seq3(new).upper())\n",
    "        cmd.get_wizard().apply()\n",
    "    print(\"\\n%s: %i peptide mutations\\n\" % (name, len(pep_mutations)))\n",
    "        \n",
    "    cmd.set_wizard()\n",
    "    #cmd.save('../sandbox/mutated/mutated.pdb', '%s//%s// %s//%s// %s//%s//' % (name, amhc, name, pept, name, b2m))\n",
    "    \n",
    "    #cmd.delete(name)\n",
    "    sleep(0.5)\n",
    "\n",
    "# wrapper for mutate(..)\n",
    "def find_and_mutate(pepseq, allele, alleledict, pep2pdb, mhc2pdb, pdb_table, first_best=10):\n",
    "    chunk = mhc2pdb.loc[mhc2pdb.index.str.startswith(allele), :]\n",
    "    best_pdbs = (chunk.apply(min) + pep2pdb.loc[pepseq, :]/10).sort_values()\n",
    "    \n",
    "    counter = 0\n",
    "    for i in xrange(len(best_pdbs)):\n",
    "        try:\n",
    "            best_pdb = best_pdbs.index[i]\n",
    "            allele_list = chunk.loc[:, best_pdb]\n",
    "            full_allele = allele_list[allele_list == min(allele_list)].index[0]\n",
    "            print(full_allele)\n",
    "            alleleseq = str(alleledict[full_allele].seq)\n",
    "            print('Compability score: %f' % best_pdbs[best_pdb])\n",
    "            \n",
    "            cmd.reinitialize()    \n",
    "            cmd.load(aligned_pdb_path+'/'+best_pdb+'_aligned.pdb', best_pdb)\n",
    "            cmd.save('../sandbox/mutated/original%i.pdb' % counter, best_pdb)\n",
    "            mutate(best_pdb, alleleseq, pepseq, pdb_table)\n",
    "            cmd.save('../sandbox/mutated/mutated%i.pdb' % counter, best_pdb)\n",
    "            cmd.delete(best_pdb)\n",
    "            \n",
    "        except AssertionError:\n",
    "            print('Assertion error: %s, %s, %s' % (pepseq, allele, best_pdb))\n",
    "            continue\n",
    "    \n",
    "        else:\n",
    "            if counter == first_best-1:\n",
    "                break\n",
    "            else:\n",
    "                counter += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 337,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'SPGMVPLHI'"
      ]
     },
     "execution_count": 337,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pepseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# allele dictionary\n",
    "alleledict = Bio.SeqIO.to_dict(Bio.SeqIO.parse(hla_fasta, 'fasta'))\n",
    "for key in alleledict.keys():\n",
    "    alleledict[key.split(';')[1]] = alleledict.pop(key)\n",
    "    \n",
    "idx = 100000\n",
    "pepseq = affinity.iloc[idx, 0]\n",
    "allele = affinity.iloc[idx, 1]\n",
    "find_and_mutate(pepseq, allele, alleledict, peptides_vs_pdb, alleles_vs_pdb, iclass)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pdb</th>\n",
       "      <th>mhc_class</th>\n",
       "      <th>host</th>\n",
       "      <th>a_chain_id</th>\n",
       "      <th>a_chain_allele</th>\n",
       "      <th>b_chain_id</th>\n",
       "      <th>b_chain_allele</th>\n",
       "      <th>antigen_id</th>\n",
       "      <th>antigen_seq</th>\n",
       "      <th>antigen_len</th>\n",
       "      <th>antigen_name</th>\n",
       "      <th>antigen_organism</th>\n",
       "      <th>comments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3i6l</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*24:02</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>QFKDNVILL</td>\n",
       "      <td>9</td>\n",
       "      <td>Nucleoprotein</td>\n",
       "      <td>SARS coronavirus Tor2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5swq</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>CVNGSCFTV</td>\n",
       "      <td>9</td>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>Influenza A virus</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3x14</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>B*08:02</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>FLRGRAYGL</td>\n",
       "      <td>9</td>\n",
       "      <td>nuclear antigen EBNA-3</td>\n",
       "      <td>Human herpesvirus 4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4jqx</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>B*44:03</td>\n",
       "      <td>C</td>\n",
       "      <td>B2M</td>\n",
       "      <td>B</td>\n",
       "      <td>EECDSELEIKRY</td>\n",
       "      <td>12</td>\n",
       "      <td>Trans-activator protein BZLF1</td>\n",
       "      <td>Human herpesvirus 4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5c0d</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>AQWGPDPAAA</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    pdb mhc_class   host a_chain_id a_chain_allele b_chain_id b_chain_allele  \\\n",
       "0  3i6l         I  Human          D        A*24:02          E            B2M   \n",
       "1  5swq         I  Human          A        A*02:01          B            B2M   \n",
       "2  3x14         I  Human          A        B*08:02          B            B2M   \n",
       "3  4jqx         I  Human          A        B*44:03          C            B2M   \n",
       "4  5c0d         I  Human          A        A*02:01          B            B2M   \n",
       "\n",
       "  antigen_id   antigen_seq  antigen_len                   antigen_name  \\\n",
       "0          F     QFKDNVILL            9                  Nucleoprotein   \n",
       "1          C     CVNGSCFTV            9                  neuraminidase   \n",
       "2          C     FLRGRAYGL            9         nuclear antigen EBNA-3   \n",
       "3          B  EECDSELEIKRY           12  Trans-activator protein BZLF1   \n",
       "4          C    AQWGPDPAAA           10                            NaN   \n",
       "\n",
       "        antigen_organism comments  \n",
       "0  SARS coronavirus Tor2      NaN  \n",
       "1      Influenza A virus      NaN  \n",
       "2    Human herpesvirus 4      NaN  \n",
       "3    Human herpesvirus 4      NaN  \n",
       "4                    NaN      NaN  "
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iclass.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def set_values(matrix, atom_desc, left_bottom, right_up, bin_size):\n",
    "    aas = ['GLY', 'ALA', 'ILE', 'LEU', \n",
    "           'PRO', 'SER', 'THR', 'CYS', \n",
    "           'MET', 'ASP', 'ASN', 'GLU', \n",
    "           'GLN', 'LYS', 'ARG', 'HIS', \n",
    "           'PHE', 'TYR', 'TRP']\n",
    "    \n",
    "    x = atom_desc[5]\n",
    "    y = atom_desc[6]\n",
    "    z = atom_desc[7]\n",
    "    \n",
    "    # check if the atom is inside of the box\n",
    "    if (x-left_bottom[0] < 0.0 or y-left_bottom[1] < 0.0 or z-left_bottom[2] < 0.0):\n",
    "        return 0\n",
    "    \n",
    "    if (x-right_up[0] > 0.0 or y-right_up[1] > 0.0 or z-right_up[2] > 0.0):\n",
    "        return 0\n",
    "    \n",
    "    # compute atom index\n",
    "    x_bin = np.floor((x-left_bottom[0])/bin_size)\n",
    "    y_bin = np.floor((y-left_bottom[1])/bin_size)\n",
    "    z_bin = np.floor((z-left_bottom[2])/bin_size)\n",
    "    \n",
    "    # increase values in matrix corresponding to the atom\n",
    "    if atom_desc[0] != 'ATOM':\n",
    "        matrix['X'][x_bin, y_bin, z_bin] += 1.0\n",
    "        matrix['XXX'][x_bin, y_bin, z_bin] += 1.0\n",
    "        return 1\n",
    "    \n",
    "    if atom_desc[1] == 'CA':\n",
    "        matrix['CA'][x_bin, y_bin, z_bin] += 1.0\n",
    "        \n",
    "    if atom_desc[2] in ['C', 'N', 'O', 'S']:\n",
    "        matrix[atom_desc[2]][x_bin, y_bin, z_bin] += 1.0\n",
    "    else:\n",
    "        matrix['X'][x_bin, y_bin, z_bin] += 1.0\n",
    "    \n",
    "    if atom_desc[3] in aas:\n",
    "        matrix[atom_desc[3]][x_bin, y_bin, z_bin] += 1.0\n",
    "    else:\n",
    "        matrix['XXX'][x_bin, y_bin, z_bin] += 1.0\n",
    "        \n",
    "    if atom_desc[4] == 'H':\n",
    "        matrix['HELIX'][x_bin, y_bin, z_bin] += 1.0\n",
    "        \n",
    "    if atom_desc[4] == 'S':\n",
    "        matrix['STRAND'][x_bin, y_bin, z_bin] += 1.0\n",
    "        \n",
    "    return 2\n",
    "\n",
    "# convert complex to 4d matrix\n",
    "# name should be preloaded to pymol with cmd.load()\n",
    "def mhc2matrix(name, left_bottom=(-10.0, -10.0, -10.0), right_up=(10.0, 10.0, 10.0), bin_size=1.0):\n",
    "    x_bins = int((right_up[0]-left_bottom[0])/bin_size)\n",
    "    y_bins = int((right_up[1]-left_bottom[1])/bin_size)\n",
    "    z_bins = int((right_up[2]-left_bottom[2])/bin_size)\n",
    "    \n",
    "    layer_names = ['CA', 'C', 'N', 'O', 'S', 'X', \n",
    "                   'GLY', 'ALA', 'ILE', 'LEU', \n",
    "                   'PRO', 'SER', 'THR', 'CYS', \n",
    "                   'MET', 'ASP', 'ASN', 'GLU', \n",
    "                   'GLN', 'LYS', 'ARG', 'HIS', \n",
    "                   'PHE', 'TYR', 'TRP', 'XXX', \n",
    "                   'HELIX', 'STRAND']\n",
    "    \n",
    "    filter_num = len(layer_names)\n",
    "    m = np.zeros((x_bins, y_bins, z_bins), dtype=zip(layer_names, [np.float32]*len(layer_names)))\n",
    "    \n",
    "    atoms = []\n",
    "    myspace = {'atoms': atoms}\n",
    "    cmd.iterate_state(0, name, 'atoms.append([type, name, elem, resn, ss, x, y, z])', space=myspace)\n",
    "    #print(atoms)\n",
    "    \n",
    "    for atom_desc in atoms: #cmd.iterate_state(0, name+'//C//', 'type, name, elem, resn, ss, x, y, z'):\n",
    "        set_values(m, atom_desc, left_bottom, right_up, bin_size)\n",
    "        \n",
    "    return np.array([m[name] for name in layer_names])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda env:python2]",
   "language": "python",
   "name": "conda-env-python2-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
