{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import Bio.PDB\n",
    "import Bio.SeqIO\n",
    "from Bio import pairwise2\n",
    "from Bio.SubsMat import MatrixInfo as matlist\n",
    "from Bio.SeqUtils import seq3\n",
    "\n",
    "import re\n",
    "import glob\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import subprocess\n",
    "\n",
    "import sys\n",
    "reload(sys)\n",
    "sys.setdefaultencoding('utf-8')\n",
    "\n",
    "# paths\n",
    "\n",
    "dbfiles = glob.glob(\"../data/mhc_alleles/hla/*prot.fasta\") + glob.glob(\"../data/mhc_alleles/mhc/*/*prot.fasta\")\n",
    "hla_fasta = '../data/mhc_alleles/hla/hla_prot.fasta'\n",
    "mhc_db_path = '../sandbox/db.fasta'\n",
    "epitope_db_path = '../data/epitopes/epitope_table.csv'\n",
    "blast_output_path = '../sandbox/blast_output'\n",
    "epitope_annot = '../sandbox/epitope_annot'\n",
    "mhc_annot = '../sandbox/mhc_annot'\n",
    "query_path = '../sandbox/query.fasta'\n",
    "pdb_path = '../data/structures'\n",
    "epitope_query_path = '../sandbox/epitope_query.fasta'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import __main__\n",
    "__main__.pymol_argv = ['pymol','-Qc'] # Pymol: quiet and no GUI\n",
    "from time import sleep\n",
    "import pymol\n",
    "from pymol import cmd\n",
    "\n",
    "pymol.finish_launching()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# replace ' ' in indentifiers with ';'\n",
    "db = open(mhc_db_path, 'w')\n",
    "for dbfile in dbfiles:\n",
    "    f = open(dbfile, 'r')\n",
    "    text = f.read()\n",
    "    f.close()\n",
    "    db.write(';'.join(text.split(' ')))\n",
    "\n",
    "db.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# fetch all chains from pdbs, combine in one query file and blast against hla db\n",
    "\n",
    "def blast_pdb():\n",
    "    parser = Bio.PDB.PDBParser()\n",
    "    ppb = Bio.PDB.CaPPBuilder()\n",
    "\n",
    "    query_file = open(query_path, 'w')\n",
    "\n",
    "    pdbfiles = glob.glob(pdb_path + \"/*.pdb\")\n",
    "    for pdb in pdbfiles:\n",
    "        structure = parser.get_structure(id=pdb[-8:-4], file=pdb)\n",
    "        for chain in structure[0]:\n",
    "            try:\n",
    "                query_file.write('>%s;%s\\n%s\\n' % (structure.id, chain.id, ''.join([str(s.get_sequence()) for s in ppb.build_peptides(chain)])))\n",
    "            except Exception:\n",
    "                continue\n",
    "\n",
    "    query_file.close()\n",
    "    \n",
    "    # submit query file for blast search\n",
    "    command = [\"blastp\", \"-query\", query_path, \n",
    "                     \"-subject\", mhc_db_path, \"-out\", \n",
    "                     blast_output_path, \"-outfmt\", str(6), \"-num_alignments\", str(1)]\n",
    "    if subprocess.call(command) != 0:\n",
    "        print('Failed to blast')\n",
    "        exit()\n",
    "    \n",
    "#blast_pdb()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# convert blast output to mhc table\n",
    "\n",
    "def host_n_allele(s):\n",
    "    host = s\n",
    "    if ';' in s:\n",
    "            host = s.split(';')[1]\n",
    "    if '-' in host:\n",
    "        return host.split('-')\n",
    "    else:\n",
    "        return ['Human', host]\n",
    "\n",
    "def make_mhc_table():\n",
    "    blasted = pd.read_csv(blast_output_path, sep='\\t', header=None)\n",
    "    pdb_list = list(set([x.split(';')[0] for x in blasted[0]]))\n",
    "    \n",
    "    mhc_table = []\n",
    "\n",
    "    for pdb in pdb_list:\n",
    "        piece = blasted[blasted[0].str.startswith(pdb)]\n",
    "        if piece.shape[0] < 2:\n",
    "            continue\n",
    "        thres = np.mean(piece[11])\n",
    "        chunk = piece[piece[11] > thres]\n",
    "        for i in chunk.index:\n",
    "            line = chunk.loc[i, :]\n",
    "            if line[2] < 90:\n",
    "                continue\n",
    "            hna = host_n_allele(line[1])\n",
    "            susp = 0\n",
    "            if line[11] < thres or line[2] < 95:\n",
    "                susp = 1\n",
    "            row = [line[0].split(';')[0], line[0].split(';')[1], 'class', hna[0], hna[1], susp]\n",
    "            mhc_table.append(row)\n",
    "\n",
    "    mhc_table = pd.DataFrame(mhc_table, columns=['pdb', 'chain', 'mhc_class', 'mhc_host', 'mhc_allele', 'unsure'])\n",
    "    mhc_table.to_csv(mhc_annot, sep='\\t', index=False)\n",
    "    \n",
    "#make_mhc_table()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# make epitope table using epitope database\n",
    "\n",
    "def make_epitope_table():\n",
    "    epitope_db = pd.read_csv(epitope_db_path, header=None, skiprows=1, low_memory=False)\n",
    "    epitope_table = []\n",
    "\n",
    "    query_file = open(query_path, 'r')\n",
    "    text = query_file.read()\n",
    "    query_file.close()\n",
    "    text = text.split()\n",
    "\n",
    "    epitope_query_file = open(epitope_query_path, 'w')\n",
    "    for i in range(len(text)):\n",
    "        if text[i][0] == '>':\n",
    "            if len(text[i+1]) < 30:\n",
    "                seq = text[i+1]\n",
    "                name = text[i].split(';')\n",
    "                epitope_query_file.write('%s\\n%s\\n' % (text[i], seq))\n",
    "                epitope_db_tmp = epitope_db[epitope_db[2].apply(lambda x: len(x)) == len(seq)]\n",
    "                epitope_db_tmp = epitope_db_tmp[epitope_db_tmp[2].str.contains(seq)]\n",
    "                trial = epitope_db_tmp[epitope_db_tmp[2] == seq]\n",
    "                if trial.shape[0] > 0:\n",
    "                    epitope_db_tmp = trial\n",
    "                if epitope_db_tmp.shape[0] > 0:\n",
    "                    row = [name[0][1:], name[1], seq, len(seq), epitope_db_tmp.iloc[0, 7], \n",
    "                           epitope_db_tmp.iloc[0, 9], epitope_db_tmp.iloc[0, 11]]\n",
    "                else:\n",
    "                    row = [name[0][1:], name[1], seq, len(seq), 'NaN', 'NaN', 'NaN']\n",
    "                epitope_table.append(row)\n",
    "\n",
    "    epitope_query_file.close()\n",
    "\n",
    "    epitope_table = pd.DataFrame(epitope_table, columns=['pdb', 'chain', 'sequence', 'length', 'name', 'organism', 'comments'])\n",
    "    epitope_table.to_csv(epitope_annot, sep='\\t', index=False)\n",
    "\n",
    "#make_epitope_table()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "epitope_table = pd.read_csv(epitope_annot, sep='\\t')\n",
    "mhc_table = pd.read_csv(mhc_annot, sep='\\t')\n",
    "\n",
    "mhc_table = mhc_table[mhc_table.mhc_host == 'Human']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# make detailed mhc table\n",
    "\n",
    "# check if chains are less than 8 A separated\n",
    "def chain_close(a, b):\n",
    "    mindist = 100\n",
    "    for a1 in a.get_atoms():\n",
    "        if mindist < 8:\n",
    "            return True\n",
    "        for a2 in b.get_atoms():\n",
    "            dist = a1-a2\n",
    "            if dist < mindist:\n",
    "                mindist = dist\n",
    "    return False\n",
    "\n",
    "# mean distance\n",
    "def chain_dist(a, b):\n",
    "    ave = 0\n",
    "    num = 0\n",
    "    for a1 in a.get_atoms():\n",
    "        mindist = 100\n",
    "        for a2 in b.get_atoms():\n",
    "            dist = a1-a2\n",
    "            if dist < mindist:\n",
    "                mindist = dist\n",
    "        ave += mindist\n",
    "        num += 1\n",
    "    return ave/num\n",
    "\n",
    "# pair MHC chains based on allele names and interchain distances\n",
    "def pair_chains(mhc, epi, alleles, pdb):\n",
    "    model = parser.get_structure(id=pdb[-8:-4], file=pdb)[0]\n",
    "    out = []\n",
    "    \n",
    "    for epi_ch in epi:\n",
    "        firmindist = 100\n",
    "        first_id = first_al = ''\n",
    "        for mhc_ch, allele in zip(mhc, alleles):\n",
    "            dist = chain_dist(model[epi_ch], model[mhc_ch])\n",
    "            if dist < firmindist:\n",
    "                firmindist = dist\n",
    "                first_id = mhc_ch\n",
    "                first_al = allele\n",
    "        \n",
    "        if len(allele.split('*')[0]) > 2:\n",
    "            second_id = second_al = ''\n",
    "            secmindist = 100\n",
    "            for mhc_ch, allele in zip(mhc, alleles):\n",
    "                if first_id == mhc_ch: \n",
    "                    continue\n",
    "                dist = chain_dist(model[epi_ch], model[mhc_ch])\n",
    "                if dist < secmindist:\n",
    "                    secmindist = dist\n",
    "                    second_id = mhc_ch\n",
    "                    second_al = allele\n",
    "            if (firmindist < 8 and secmindist < 8 and chain_close(model[first_id], model[second_id])):\n",
    "                if (first_al.split('*')[0][2] == 'B' and second_al.split('*')[0][2] == 'A'):\n",
    "                    out.append([[second_id, first_id], epi_ch])\n",
    "                else:\n",
    "                    out.append([[first_id, second_id], epi_ch])\n",
    "        else:\n",
    "            out.append([[first_id], epi_ch])\n",
    "    return out\n",
    "\n",
    "# verify intersection absence\n",
    "def correctness(pairs):\n",
    "    for i in range(len(pairs)-1):\n",
    "        for j in range(i+1, len(pairs)):\n",
    "            for l in range(len(pairs[i][0])):\n",
    "                for m in range(len(pairs[j][0])):\n",
    "                    assert(pairs[i][0][l] != pairs[j][0][m])\n",
    "\n",
    "\n",
    "def make_annotation_table():\n",
    "    parser = Bio.PDB.PDBParser(QUIET=True)\n",
    "    ppb = Bio.PDB.CaPPBuilder()                    \n",
    "    pdb_set = list(set(mhc_table.pdb))\n",
    "    annotation = []\n",
    "\n",
    "    for pdb in pdb_set[277:]:\n",
    "        mhc_chunk = mhc_table[mhc_table.pdb == pdb]\n",
    "        epi_chunk = epitope_table[epitope_table.pdb == pdb]\n",
    "        pairs = pair_chains(list(mhc_chunk.chain), list(epi_chunk.chain), list(mhc_chunk.mhc_allele), \n",
    "                            \"%s/%s.pdb\" % (pdb_path, pdb))\n",
    "        correctness(pairs)\n",
    "\n",
    "        for pair in pairs:\n",
    "            line = ''\n",
    "            if len(pair[0]) == 1:\n",
    "                mhc_local = mhc_chunk[mhc_chunk.chain == pair[0][0]]\n",
    "                assert(mhc_local.iloc[0, 4][1] == '*')\n",
    "                #               pdb          class         host              1 chain  \n",
    "                line = [mhc_local.iloc[0, 0], 'I', mhc_local.iloc[0, 3],  mhc_local.iloc[0, 1], \n",
    "                #         1 chain allele       2 chain   2 chain allele      \n",
    "                        mhc_local.iloc[0, 4],   '',         'B2M']\n",
    "\n",
    "            if len(pair[0]) == 2:\n",
    "                mhc_a = mhc_chunk[mhc_chunk.chain == pair[0][0]]\n",
    "                mhc_b = mhc_chunk[mhc_chunk.chain == pair[0][1]]\n",
    "\n",
    "                assert(mhc_a.iloc[0, 4][:2] == mhc_b.iloc[0, 4][:2])\n",
    "                assert(mhc_a.iloc[0, 4][2] == 'A')\n",
    "                assert(mhc_b.iloc[0, 4][2] == 'B')\n",
    "                line = [mhc_chunk.iloc[0, 0], 'II', mhc_chunk.iloc[0, 3], mhc_a.iloc[0, 1], \n",
    "                        mhc_a.iloc[0, 4], mhc_b.iloc[0, 1], mhc_b.iloc[0, 4]]\n",
    "\n",
    "            line = line + list(epi_chunk[epi_chunk.chain == pair[1]].iloc[0, 1:])\n",
    "        annotation.append(line)\n",
    "\n",
    "        cols = ['pdb', 'mhc_class', 'host', 'a_chain_id', 'a_chain_allele', 'b_chain_id', 'b_chain_allele', \n",
    "            'antigen_id', 'antigen_seq', 'antigen_len', 'antigen_name', 'antigen_organism', 'comments']\n",
    "        annotation = pd.DataFrame(annotation, columns=cols).drop_duplicates()\n",
    "        annotation.to_csv('../sandbox/detailed_mhc', index=None, sep='\\t')\n",
    "        \n",
    "#make_annotation_table()\n",
    "\n",
    "annotation = pd.read_csv('../sandbox/detailed_mhc', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# define microglobuline chains\n",
    "def b2m_align(s):\n",
    "    b2m = 'MSRSVALAVLALLSLSGLEAIQRTPKIQVYSRHPAENGKSNFLNCYVSGFHPSDIEVDLLKNGERIEKDEYACRVNHVTLSQPKIVKWDRDM'\n",
    "    matrix = matlist.blosum90\n",
    "    gap_open = -1.0\n",
    "    gap_extend = -0.5\n",
    "    alns = pairwise2.align.globalds(s, b2m, matrix, gap_open, gap_extend)\n",
    "    return alns[0]\n",
    "\n",
    "def b2m_define(table):\n",
    "    chunk = table[table.mhc_class == 'I']\n",
    "    parser = Bio.PDB.PDBParser(QUIET=True)\n",
    "    ppb = Bio.PDB.CaPPBuilder() \n",
    "    \n",
    "    for i in chunk.index:\n",
    "        pdb = chunk.loc[i, 'pdb']\n",
    "        print(pdb)\n",
    "        structure = parser.get_structure(id=pdb, file=pdb_path+'/'+pdb+'.pdb')[0]\n",
    "        mhc = chunk.loc[i, 'a_chain_id']\n",
    "        pep = chunk.loc[i, 'antigen_id']\n",
    "        mindist = 1000\n",
    "        b2m = None;\n",
    "        for chain in structure.get_chains():\n",
    "            if chain.get_id() != mhc and chain.get_id() != pep:\n",
    "                if len(list(structure.get_chains())) == 3:\n",
    "                    b2m = chain\n",
    "                else:\n",
    "                    chain_seq = ''.join([str(x.get_sequence()) for x in ppb.build_peptides(chain)])\n",
    "                    if b2m_align(chain_seq)[2] > 400:\n",
    "                        if chain_close(chain, structure[mhc]):\n",
    "                            dist = chain_dist(chain, structure[mhc])\n",
    "                            if dist <= mindist:\n",
    "                                mindist = dist\n",
    "                                b2m = chain\n",
    "                                \n",
    "        if b2m == None:\n",
    "            table.loc[i, 'b_chain_id'] = None\n",
    "            print('bsm chain = %s' % 'None')\n",
    "        else:\n",
    "            table.loc[i, 'b_chain_id'] = b2m.get_id()\n",
    "            print('bsm chain = %s' % b2m.get_id())\n",
    "\n",
    "    annotation.to_csv('../sandbox/detailed_mhc', index=None, sep='\\t')\n",
    "        \n",
    "#b2m_define(annotation)\n",
    "\n",
    "annotation = pd.read_csv('../sandbox/detailed_mhc', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "iclass = annotation[annotation.mhc_class == 'I']\n",
    "iclass = iclass[iclass.a_chain_allele.str.contains('[ABC]\\*', regex=True)]\n",
    "iclass.a_chain_allele = [':'.join(x.split(':')[:2]) for x in iclass.a_chain_allele]\n",
    "iclass = iclass.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "tmp = iclass[iclass.a_chain_allele.str.startswith('A*02:01')]\n",
    "with open('../sandbox/peptides_cluster', 'w') as file:\n",
    "    for i in range(tmp.shape[0]):\n",
    "        file.write('>%s\\n%s\\n' % (tmp.iloc[i, 0], tmp.iloc[i, 8]))\n",
    "        \n",
    "\n",
    "cluster1 = ['2x4r', '3gsv', '3gsq', '3mrd', '3mrb', '3mrc', '3gsx', '3gsw', '3gsu', '3gsr', '3mr9']\n",
    "cluster2 = ['4l3c', '1s9w', '3kla', '3gjf', '1s9y', '1s9x']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def align_pdb(table):\n",
    "    \n",
    "    class SelectChains(Bio.PDB.Select):\n",
    "        def __init__(self, chain_letters):\n",
    "            self.chain_letters = chain_letters\n",
    "        def accept_chain(self, chain):\n",
    "            return (chain.get_id() in self.chain_letters)\n",
    "\n",
    "    parser = Bio.PDB.PDBParser(QUIET=True)\n",
    "\n",
    "    name = table.iloc[0, 0]; amhc = table.iloc[0, 3]; pept = table.iloc[0, 7]; b2m = table.iloc[0, 5]\n",
    "    ref = parser.get_structure(id=name, file=pdb_path+'/'+name+'.pdb')[0]\n",
    "    ref_atoms = []\n",
    "    for atom in ref[amhc].get_atoms():\n",
    "        if atom.get_name() == 'CA':\n",
    "            ref_atoms.append(atom)\n",
    "\n",
    "    io = Bio.PDB.PDBIO()\n",
    "    io.set_structure(ref)\n",
    "    io.save('../sandbox/aligned_pdb/%s_cut.pdb' % name, SelectChains([amhc, pept, b2m]))\n",
    "\n",
    "    super_imposer = Bio.PDB.Superimposer()\n",
    "\n",
    "    for i in range(1, table.shape[0]):\n",
    "        name = table.iloc[i, 0]\n",
    "        amhc = table.iloc[i, 3]\n",
    "        b2m  = table.iloc[i, 5]\n",
    "        pept = table.iloc[i, 7]\n",
    "        sample = parser.get_structure(id=name, file=pdb_path+'/'+name+'.pdb')[0]\n",
    "        sample_atoms = []\n",
    "        for atom in sample[amhc].get_atoms():\n",
    "            if atom.get_name() == 'CA':\n",
    "                sample_atoms.append(atom)\n",
    "                \n",
    "        length = min(len(ref_atoms), len(sample_atoms))\n",
    "        super_imposer.set_atoms(ref_atoms[:length], sample_atoms[:length])\n",
    "        super_imposer.apply(sample.get_atoms())\n",
    "        \n",
    "        io.set_structure(sample)\n",
    "        io.save('../sandbox/aligned_pdb/%s_cut.pdb' % name, SelectChains([amhc, pept, b2m]))\n",
    "        \n",
    "align_pdb(iclass[iclass.pdb.isin(['3mrb', '4i48'])])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pdb</th>\n",
       "      <th>mhc_class</th>\n",
       "      <th>host</th>\n",
       "      <th>a_chain_id</th>\n",
       "      <th>a_chain_allele</th>\n",
       "      <th>b_chain_id</th>\n",
       "      <th>b_chain_allele</th>\n",
       "      <th>antigen_id</th>\n",
       "      <th>antigen_seq</th>\n",
       "      <th>antigen_len</th>\n",
       "      <th>antigen_name</th>\n",
       "      <th>antigen_organism</th>\n",
       "      <th>comments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3i6l</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*24:02</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>QFKDNVILL</td>\n",
       "      <td>9</td>\n",
       "      <td>Nucleoprotein</td>\n",
       "      <td>SARS coronavirus Tor2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5swq</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>CVNGSCFTV</td>\n",
       "      <td>9</td>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>Influenza A virus</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>3mri</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>P</td>\n",
       "      <td>CINMWCWTV</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2uwe</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>H</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>I</td>\n",
       "      <td>B2M</td>\n",
       "      <td>J</td>\n",
       "      <td>ALWGFFPVL</td>\n",
       "      <td>9</td>\n",
       "      <td>chromosome 15 open reading frame 24</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>3gsw</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>P</td>\n",
       "      <td>NLVPMVAAV</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>3hpj</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>RMFPNAPYL</td>\n",
       "      <td>9</td>\n",
       "      <td>Wilms tumor protein</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2git</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>LLFGKPVYV</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>3qeq</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>AAGIGILTV</td>\n",
       "      <td>9</td>\n",
       "      <td>Melanoma antigen recognized by T-cells 1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>4hx1</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*68:02</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>SVYDFFVWL</td>\n",
       "      <td>9</td>\n",
       "      <td>tyrosinase-related protein-2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1b0g</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>ALWGFFPVL</td>\n",
       "      <td>9</td>\n",
       "      <td>chromosome 15 open reading frame 24</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1ao7</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>LLFGYPVYV</td>\n",
       "      <td>9</td>\n",
       "      <td>Protein Tax-1</td>\n",
       "      <td>Human T-lymphotropic virus 1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>2jcc</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>H</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>I</td>\n",
       "      <td>B2M</td>\n",
       "      <td>J</td>\n",
       "      <td>ALWGFFPVL</td>\n",
       "      <td>9</td>\n",
       "      <td>chromosome 15 open reading frame 24</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>3bh9</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>RTYGPMNKV</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>1q94</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*11:01</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>AIFQSSMTK</td>\n",
       "      <td>9</td>\n",
       "      <td>Pol polyprotein</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>2gtz</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>ALGIGILTV</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>The epitope is an analog of a MART-1/Melan-A 2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>2gtw</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>LAGIGILTV</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>The epitope is an analog of a MART-1/Melan-A 2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>3mrb</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>P</td>\n",
       "      <td>NLVPMVHTV</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>4nqx</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>K</td>\n",
       "      <td>A*01:01</td>\n",
       "      <td>L</td>\n",
       "      <td>B2M</td>\n",
       "      <td>R</td>\n",
       "      <td>CTELKLNDY</td>\n",
       "      <td>9</td>\n",
       "      <td>nucleocapsid protein</td>\n",
       "      <td>H1N1 subtype</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>1t21</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>SLYNTVATL</td>\n",
       "      <td>9</td>\n",
       "      <td>gag-pol polyprotein</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>4n8v</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*11:01</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>MLIYSMWGK</td>\n",
       "      <td>9</td>\n",
       "      <td>Virion envelope protein A14</td>\n",
       "      <td>Vaccinia virus Copenhagen</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>3qdj</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>AAGIGILTV</td>\n",
       "      <td>9</td>\n",
       "      <td>Melanoma antigen recognized by T-cells 1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>1akj</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>ILKEPVHGV</td>\n",
       "      <td>9</td>\n",
       "      <td>Gag-Pol polyprotein</td>\n",
       "      <td>Human immunodeficiency virus type 1 BH10</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>4i48</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*68:02</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>TLTSCNTSV</td>\n",
       "      <td>9</td>\n",
       "      <td>protein env-lor</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>2vlr</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>F</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>G</td>\n",
       "      <td>B2M</td>\n",
       "      <td>H</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "      <td>9</td>\n",
       "      <td>matrix protein</td>\n",
       "      <td>Influenza A virus</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>1eey</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>ILSALVGIV</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>The epitope is a GP2 variant with mutations: I...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>1hhj</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>ILKEPVHGV</td>\n",
       "      <td>9</td>\n",
       "      <td>Gag-Pol polyprotein</td>\n",
       "      <td>Human immunodeficiency virus type 1 BH10</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>2x4r</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>NLVPMVATV</td>\n",
       "      <td>9</td>\n",
       "      <td>pp65</td>\n",
       "      <td>Human herpesvirus 5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>5eu5</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>YLEPAPVTA</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>1hhk</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>LLFGYPVYV</td>\n",
       "      <td>9</td>\n",
       "      <td>Protein Tax-1</td>\n",
       "      <td>Human T-lymphotropic virus 1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>1s9y</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>SLLMWITQS</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>This peptide is an analog of cancer/testis ant...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336</th>\n",
       "      <td>1hhi</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "      <td>9</td>\n",
       "      <td>matrix protein</td>\n",
       "      <td>Influenza A virus</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339</th>\n",
       "      <td>5eu3</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>YLEPGPVTA</td>\n",
       "      <td>9</td>\n",
       "      <td>Melanocyte protein Pmel 17 precursor</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>3gsv</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>P</td>\n",
       "      <td>NLVPQVATV</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>346</th>\n",
       "      <td>1s9x</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>SLLMWITQA</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>This peptide is an analog of cancer/testis ant...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>348</th>\n",
       "      <td>2x4u</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>ILKEPVHGV</td>\n",
       "      <td>9</td>\n",
       "      <td>Gag-Pol polyprotein</td>\n",
       "      <td>Human immunodeficiency virus type 1 BH10</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>3bhb</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>KMDFLDMQL</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>352</th>\n",
       "      <td>2bnq</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>SLLMWITQV</td>\n",
       "      <td>9</td>\n",
       "      <td>Chain C, Structural And Kinetic Basis For Heig...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>5eu4</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>YLAPGPVTA</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Imported from the TopBank database. This pepti...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>362</th>\n",
       "      <td>1t1x</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>SLYLTVATL</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>366</th>\n",
       "      <td>2f54</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>F</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>G</td>\n",
       "      <td>B2M</td>\n",
       "      <td>H</td>\n",
       "      <td>SLLMWITQC</td>\n",
       "      <td>9</td>\n",
       "      <td>Cancer/testis antigen 1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369</th>\n",
       "      <td>2pye</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>SLLMWITQC</td>\n",
       "      <td>9</td>\n",
       "      <td>Cancer/testis antigen 1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>370</th>\n",
       "      <td>2v2w</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>SLYNTVATL</td>\n",
       "      <td>9</td>\n",
       "      <td>gag-pol polyprotein</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>371</th>\n",
       "      <td>3mrd</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>P</td>\n",
       "      <td>NLVPMGATV</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374</th>\n",
       "      <td>3gso</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>P</td>\n",
       "      <td>NLVPMVATV</td>\n",
       "      <td>9</td>\n",
       "      <td>pp65</td>\n",
       "      <td>Human herpesvirus 5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377</th>\n",
       "      <td>2c7u</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>SLFNTIAVL</td>\n",
       "      <td>9</td>\n",
       "      <td>gag protein</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>383</th>\n",
       "      <td>1im3</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>M</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>N</td>\n",
       "      <td>B2M</td>\n",
       "      <td>O</td>\n",
       "      <td>LLFGYPVYV</td>\n",
       "      <td>9</td>\n",
       "      <td>Protein Tax-1</td>\n",
       "      <td>Human T-lymphotropic virus 1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>385</th>\n",
       "      <td>3pwj</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>LLYGFVNYV</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387</th>\n",
       "      <td>4hwz</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*68:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>AIFQSSMTK</td>\n",
       "      <td>9</td>\n",
       "      <td>Pol polyprotein</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>390</th>\n",
       "      <td>3d25</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>VLHDDLLEA</td>\n",
       "      <td>9</td>\n",
       "      <td>histocompatibility (minor) HA-1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>393</th>\n",
       "      <td>3rew</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>CLGGLLTMV</td>\n",
       "      <td>9</td>\n",
       "      <td>terminal protein LMP2B</td>\n",
       "      <td>Human herpesvirus 4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394</th>\n",
       "      <td>2p5e</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>SLLMWITQC</td>\n",
       "      <td>9</td>\n",
       "      <td>Cancer/testis antigen 1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>399</th>\n",
       "      <td>4e5x</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>LLFGYPVYV</td>\n",
       "      <td>9</td>\n",
       "      <td>Protein Tax-1</td>\n",
       "      <td>Human T-lymphotropic virus 1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>402</th>\n",
       "      <td>1duz</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>LLFGYPVYV</td>\n",
       "      <td>9</td>\n",
       "      <td>Protein Tax-1</td>\n",
       "      <td>Human T-lymphotropic virus 1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>403</th>\n",
       "      <td>1oga</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "      <td>9</td>\n",
       "      <td>matrix protein</td>\n",
       "      <td>Influenza A virus</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>3h7b</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>MLWGYLQYV</td>\n",
       "      <td>9</td>\n",
       "      <td>telomer length regulation protein TEL1</td>\n",
       "      <td>Saccharomyces cerevisiae RM11-1a</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>1jht</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>ALGIGILTV</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>The epitope is an analog of a MART-1/Melan-A 2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>415</th>\n",
       "      <td>4i4w</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>ILAKFLHRL</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>425</th>\n",
       "      <td>5eu6</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>YLEPGPVTV</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>427</th>\n",
       "      <td>3mrj</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>P</td>\n",
       "      <td>CINGMCWTV</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>430</th>\n",
       "      <td>1tvh</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>IMDQVPFSV</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>The epitope is an analog of the gp100(209-217)...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>155 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      pdb mhc_class   host a_chain_id a_chain_allele b_chain_id  \\\n",
       "0    3i6l         I  Human          D        A*24:02          E   \n",
       "1    5swq         I  Human          A        A*02:01          B   \n",
       "5    3mri         I  Human          A        A*02:01          B   \n",
       "7    2uwe         I  Human          H        A*02:01          I   \n",
       "8    3gsw         I  Human          A        A*02:01          B   \n",
       "12   3hpj         I  Human          D        A*02:01          E   \n",
       "13   2git         I  Human          D        A*02:01          E   \n",
       "21   3qeq         I  Human          A        A*02:01          B   \n",
       "22   4hx1         I  Human          A        A*68:02          B   \n",
       "23   1b0g         I  Human          D        A*02:01          E   \n",
       "29   1ao7         I  Human          A        A*02:01          B   \n",
       "30   2jcc         I  Human          H        A*02:01          I   \n",
       "42   3bh9         I  Human          A        A*02:01          B   \n",
       "45   1q94         I  Human          D        A*11:01          E   \n",
       "46   2gtz         I  Human          D        A*02:01          E   \n",
       "48   2gtw         I  Human          D        A*02:01          E   \n",
       "53   3mrb         I  Human          A        A*02:01          B   \n",
       "56   4nqx         I  Human          K        A*01:01          L   \n",
       "63   1t21         I  Human          A        A*02:01          B   \n",
       "64   4n8v         I  Human          D        A*11:01          E   \n",
       "75   3qdj         I  Human          A        A*02:01          B   \n",
       "77   1akj         I  Human          A        A*02:01          B   \n",
       "78   4i48         I  Human          A        A*68:02          B   \n",
       "80   2vlr         I  Human          F        A*02:01          G   \n",
       "81   1eey         I  Human          D        A*02:01          E   \n",
       "82   1hhj         I  Human          D        A*02:01          E   \n",
       "85   2x4r         I  Human          D        A*02:01          E   \n",
       "88   5eu5         I  Human          A        A*02:01          B   \n",
       "90   1hhk         I  Human          D        A*02:01          E   \n",
       "91   1s9y         I  Human          A        A*02:01          B   \n",
       "..    ...       ...    ...        ...            ...        ...   \n",
       "336  1hhi         I  Human          D        A*02:01          E   \n",
       "339  5eu3         I  Human          A        A*02:01          B   \n",
       "345  3gsv         I  Human          A        A*02:01          B   \n",
       "346  1s9x         I  Human          A        A*02:01          B   \n",
       "348  2x4u         I  Human          D        A*02:01          E   \n",
       "350  3bhb         I  Human          A        A*02:01          B   \n",
       "352  2bnq         I  Human          A        A*02:01          B   \n",
       "355  5eu4         I  Human          D        A*02:01          E   \n",
       "362  1t1x         I  Human          A        A*02:01          B   \n",
       "366  2f54         I  Human          F        A*02:01          G   \n",
       "369  2pye         I  Human          A        A*02:01          B   \n",
       "370  2v2w         I  Human          D        A*02:01          E   \n",
       "371  3mrd         I  Human          A        A*02:01          B   \n",
       "374  3gso         I  Human          A        A*02:01          B   \n",
       "377  2c7u         I  Human          D        A*02:01          E   \n",
       "383  1im3         I  Human          M        A*02:01          N   \n",
       "385  3pwj         I  Human          D        A*02:01          E   \n",
       "387  4hwz         I  Human          A        A*68:01          B   \n",
       "390  3d25         I  Human          A        A*02:01          B   \n",
       "393  3rew         I  Human          D        A*02:01          E   \n",
       "394  2p5e         I  Human          A        A*02:01          B   \n",
       "399  4e5x         I  Human          D        A*02:01          E   \n",
       "402  1duz         I  Human          D        A*02:01          E   \n",
       "403  1oga         I  Human          A        A*02:01          B   \n",
       "406  3h7b         I  Human          D        A*02:01          E   \n",
       "407  1jht         I  Human          A        A*02:01          B   \n",
       "415  4i4w         I  Human          A        A*02:01          B   \n",
       "425  5eu6         I  Human          A        A*02:01          B   \n",
       "427  3mrj         I  Human          A        A*02:01          B   \n",
       "430  1tvh         I  Human          D        A*02:01          E   \n",
       "\n",
       "    b_chain_allele antigen_id antigen_seq  antigen_len  \\\n",
       "0              B2M          F   QFKDNVILL            9   \n",
       "1              B2M          C   CVNGSCFTV            9   \n",
       "5              B2M          P   CINMWCWTV            9   \n",
       "7              B2M          J   ALWGFFPVL            9   \n",
       "8              B2M          P   NLVPMVAAV            9   \n",
       "12             B2M          F   RMFPNAPYL            9   \n",
       "13             B2M          F   LLFGKPVYV            9   \n",
       "21             B2M          C   AAGIGILTV            9   \n",
       "22             B2M          C   SVYDFFVWL            9   \n",
       "23             B2M          F   ALWGFFPVL            9   \n",
       "29             B2M          C   LLFGYPVYV            9   \n",
       "30             B2M          J   ALWGFFPVL            9   \n",
       "42             B2M          C   RTYGPMNKV            9   \n",
       "45             B2M          F   AIFQSSMTK            9   \n",
       "46             B2M          F   ALGIGILTV            9   \n",
       "48             B2M          F   LAGIGILTV            9   \n",
       "53             B2M          P   NLVPMVHTV            9   \n",
       "56             B2M          R   CTELKLNDY            9   \n",
       "63             B2M          C   SLYNTVATL            9   \n",
       "64             B2M          F   MLIYSMWGK            9   \n",
       "75             B2M          C   AAGIGILTV            9   \n",
       "77             B2M          C   ILKEPVHGV            9   \n",
       "78             B2M          C   TLTSCNTSV            9   \n",
       "80             B2M          H   GILGFVFTL            9   \n",
       "81             B2M          F   ILSALVGIV            9   \n",
       "82             B2M          F   ILKEPVHGV            9   \n",
       "85             B2M          F   NLVPMVATV            9   \n",
       "88             B2M          C   YLEPAPVTA            9   \n",
       "90             B2M          F   LLFGYPVYV            9   \n",
       "91             B2M          C   SLLMWITQS            9   \n",
       "..             ...        ...         ...          ...   \n",
       "336            B2M          F   GILGFVFTL            9   \n",
       "339            B2M          C   YLEPGPVTA            9   \n",
       "345            B2M          P   NLVPQVATV            9   \n",
       "346            B2M          C   SLLMWITQA            9   \n",
       "348            B2M          F   ILKEPVHGV            9   \n",
       "350            B2M          C   KMDFLDMQL            9   \n",
       "352            B2M          C   SLLMWITQV            9   \n",
       "355            B2M          F   YLAPGPVTA            9   \n",
       "362            B2M          C   SLYLTVATL            9   \n",
       "366            B2M          H   SLLMWITQC            9   \n",
       "369            B2M          C   SLLMWITQC            9   \n",
       "370            B2M          F   SLYNTVATL            9   \n",
       "371            B2M          P   NLVPMGATV            9   \n",
       "374            B2M          P   NLVPMVATV            9   \n",
       "377            B2M          F   SLFNTIAVL            9   \n",
       "383            B2M          O   LLFGYPVYV            9   \n",
       "385            B2M          F   LLYGFVNYV            9   \n",
       "387            B2M          C   AIFQSSMTK            9   \n",
       "390            B2M          C   VLHDDLLEA            9   \n",
       "393            B2M          F   CLGGLLTMV            9   \n",
       "394            B2M          C   SLLMWITQC            9   \n",
       "399            B2M          F   LLFGYPVYV            9   \n",
       "402            B2M          F   LLFGYPVYV            9   \n",
       "403            B2M          C   GILGFVFTL            9   \n",
       "406            B2M          F   MLWGYLQYV            9   \n",
       "407            B2M          C   ALGIGILTV            9   \n",
       "415            B2M          C   ILAKFLHRL            9   \n",
       "425            B2M          C   YLEPGPVTV            9   \n",
       "427            B2M          P   CINGMCWTV            9   \n",
       "430            B2M          F   IMDQVPFSV            9   \n",
       "\n",
       "                                          antigen_name  \\\n",
       "0                                        Nucleoprotein   \n",
       "1                                        neuraminidase   \n",
       "5                                                  NaN   \n",
       "7                  chromosome 15 open reading frame 24   \n",
       "8                                                  NaN   \n",
       "12                                 Wilms tumor protein   \n",
       "13                                                 NaN   \n",
       "21            Melanoma antigen recognized by T-cells 1   \n",
       "22                        tyrosinase-related protein-2   \n",
       "23                 chromosome 15 open reading frame 24   \n",
       "29                                       Protein Tax-1   \n",
       "30                 chromosome 15 open reading frame 24   \n",
       "42                                                 NaN   \n",
       "45                                     Pol polyprotein   \n",
       "46                                                 NaN   \n",
       "48                                                 NaN   \n",
       "53                                                 NaN   \n",
       "56                                nucleocapsid protein   \n",
       "63                                 gag-pol polyprotein   \n",
       "64                         Virion envelope protein A14   \n",
       "75            Melanoma antigen recognized by T-cells 1   \n",
       "77                                 Gag-Pol polyprotein   \n",
       "78                                     protein env-lor   \n",
       "80                                      matrix protein   \n",
       "81                                                 NaN   \n",
       "82                                 Gag-Pol polyprotein   \n",
       "85                                                pp65   \n",
       "88                                                 NaN   \n",
       "90                                       Protein Tax-1   \n",
       "91                                                 NaN   \n",
       "..                                                 ...   \n",
       "336                                     matrix protein   \n",
       "339               Melanocyte protein Pmel 17 precursor   \n",
       "345                                                NaN   \n",
       "346                                                NaN   \n",
       "348                                Gag-Pol polyprotein   \n",
       "350                                                NaN   \n",
       "352  Chain C, Structural And Kinetic Basis For Heig...   \n",
       "355                                                NaN   \n",
       "362                                                NaN   \n",
       "366                            Cancer/testis antigen 1   \n",
       "369                            Cancer/testis antigen 1   \n",
       "370                                gag-pol polyprotein   \n",
       "371                                                NaN   \n",
       "374                                               pp65   \n",
       "377                                        gag protein   \n",
       "383                                      Protein Tax-1   \n",
       "385                                                NaN   \n",
       "387                                    Pol polyprotein   \n",
       "390                    histocompatibility (minor) HA-1   \n",
       "393                             terminal protein LMP2B   \n",
       "394                            Cancer/testis antigen 1   \n",
       "399                                      Protein Tax-1   \n",
       "402                                      Protein Tax-1   \n",
       "403                                     matrix protein   \n",
       "406             telomer length regulation protein TEL1   \n",
       "407                                                NaN   \n",
       "415                                                NaN   \n",
       "425                                                NaN   \n",
       "427                                                NaN   \n",
       "430                                                NaN   \n",
       "\n",
       "                             antigen_organism  \\\n",
       "0                       SARS coronavirus Tor2   \n",
       "1                           Influenza A virus   \n",
       "5                                         NaN   \n",
       "7                                Homo sapiens   \n",
       "8                                         NaN   \n",
       "12                               Homo sapiens   \n",
       "13                                        NaN   \n",
       "21                               Homo sapiens   \n",
       "22                               Homo sapiens   \n",
       "23                               Homo sapiens   \n",
       "29               Human T-lymphotropic virus 1   \n",
       "30                               Homo sapiens   \n",
       "42                                        NaN   \n",
       "45             Human immunodeficiency virus 1   \n",
       "46                                        NaN   \n",
       "48                                        NaN   \n",
       "53                                        NaN   \n",
       "56                               H1N1 subtype   \n",
       "63             Human immunodeficiency virus 1   \n",
       "64                  Vaccinia virus Copenhagen   \n",
       "75                               Homo sapiens   \n",
       "77   Human immunodeficiency virus type 1 BH10   \n",
       "78             Human immunodeficiency virus 1   \n",
       "80                          Influenza A virus   \n",
       "81                                        NaN   \n",
       "82   Human immunodeficiency virus type 1 BH10   \n",
       "85                        Human herpesvirus 5   \n",
       "88                                        NaN   \n",
       "90               Human T-lymphotropic virus 1   \n",
       "91                                        NaN   \n",
       "..                                        ...   \n",
       "336                         Influenza A virus   \n",
       "339                              Homo sapiens   \n",
       "345                                       NaN   \n",
       "346                                       NaN   \n",
       "348  Human immunodeficiency virus type 1 BH10   \n",
       "350                                       NaN   \n",
       "352                              Homo sapiens   \n",
       "355                                       NaN   \n",
       "362                                       NaN   \n",
       "366                              Homo sapiens   \n",
       "369                              Homo sapiens   \n",
       "370            Human immunodeficiency virus 1   \n",
       "371                                       NaN   \n",
       "374                       Human herpesvirus 5   \n",
       "377            Human immunodeficiency virus 1   \n",
       "383              Human T-lymphotropic virus 1   \n",
       "385                                       NaN   \n",
       "387            Human immunodeficiency virus 1   \n",
       "390                              Homo sapiens   \n",
       "393                       Human herpesvirus 4   \n",
       "394                              Homo sapiens   \n",
       "399              Human T-lymphotropic virus 1   \n",
       "402              Human T-lymphotropic virus 1   \n",
       "403                         Influenza A virus   \n",
       "406          Saccharomyces cerevisiae RM11-1a   \n",
       "407                                       NaN   \n",
       "415                                       NaN   \n",
       "425                                       NaN   \n",
       "427                                       NaN   \n",
       "430                                       NaN   \n",
       "\n",
       "                                              comments  \n",
       "0                                                  NaN  \n",
       "1                                                  NaN  \n",
       "5                                                  NaN  \n",
       "7                                                  NaN  \n",
       "8                                                  NaN  \n",
       "12                                                 NaN  \n",
       "13                                                 NaN  \n",
       "21                                                 NaN  \n",
       "22                                                 NaN  \n",
       "23                                                 NaN  \n",
       "29                                                 NaN  \n",
       "30                                                 NaN  \n",
       "42                                                 NaN  \n",
       "45                                                 NaN  \n",
       "46   The epitope is an analog of a MART-1/Melan-A 2...  \n",
       "48   The epitope is an analog of a MART-1/Melan-A 2...  \n",
       "53                                                 NaN  \n",
       "56                                                 NaN  \n",
       "63                                                 NaN  \n",
       "64                                                 NaN  \n",
       "75                                                 NaN  \n",
       "77                                                 NaN  \n",
       "78                                                 NaN  \n",
       "80                                                 NaN  \n",
       "81   The epitope is a GP2 variant with mutations: I...  \n",
       "82                                                 NaN  \n",
       "85                                                 NaN  \n",
       "88                                                 NaN  \n",
       "90                                                 NaN  \n",
       "91   This peptide is an analog of cancer/testis ant...  \n",
       "..                                                 ...  \n",
       "336                                                NaN  \n",
       "339                                                NaN  \n",
       "345                                                NaN  \n",
       "346  This peptide is an analog of cancer/testis ant...  \n",
       "348                                                NaN  \n",
       "350                                                NaN  \n",
       "352                                                NaN  \n",
       "355  Imported from the TopBank database. This pepti...  \n",
       "362                                                NaN  \n",
       "366                                                NaN  \n",
       "369                                                NaN  \n",
       "370                                                NaN  \n",
       "371                                                NaN  \n",
       "374                                                NaN  \n",
       "377                                                NaN  \n",
       "383                                                NaN  \n",
       "385                                                NaN  \n",
       "387                                                NaN  \n",
       "390                                                NaN  \n",
       "393                                                NaN  \n",
       "394                                                NaN  \n",
       "399                                                NaN  \n",
       "402                                                NaN  \n",
       "403                                                NaN  \n",
       "406                                                NaN  \n",
       "407  The epitope is an analog of a MART-1/Melan-A 2...  \n",
       "415                                                NaN  \n",
       "425                                                NaN  \n",
       "427                                                NaN  \n",
       "430  The epitope is an analog of the gp100(209-217)...  \n",
       "\n",
       "[155 rows x 13 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iclass[(iclass.antigen_len == 9) & (iclass.a_chain_allele.str.startswith('A'))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# if residue in mhc is close to antigen then return 1.\n",
    "# if further return 0.5\n",
    "# if far return 0.01\n",
    "def get_points(res, chain):\n",
    "    mdist = 1000\n",
    "    for a in res:\n",
    "        for b in chain.get_atoms():\n",
    "            dist = a-b\n",
    "            if dist < mdist:\n",
    "                mdist = dist\n",
    "                if mdist < 5:\n",
    "                    return 1.0\n",
    "    if mdist < 10:\n",
    "        return 0.5\n",
    "    return 0.01\n",
    "\n",
    "# return sorted list of best alleles\n",
    "def score_alignments(scores, blasted, points, alleles, line_id, size):\n",
    "    p1 = re.compile('\\w+\\s+([0-9]+)\\s+(\\w+)\\s+([0-9]+)\\\\n')\n",
    "    p2 = re.compile('Subject_([0-9]+)')\n",
    "    p3 = re.compile('[A-Z]')\n",
    "\n",
    "    res_start = start = end = counter = 0\n",
    "    \n",
    "    for line in blasted:\n",
    "        if line[:7] == 'Query_1':\n",
    "            m = p1.match(line)\n",
    "            res_start = int(m.group(1))-1\n",
    "            start = m.start(2)\n",
    "            end = m.end(2)\n",
    "            counter = 0\n",
    "            \n",
    "        if line[:8] == 'Subject_':\n",
    "            if counter >= size:\n",
    "                break\n",
    "            allele_id = int(p2.match(line).group(1))-1\n",
    "            chunk = line[start:end+1]\n",
    "            if np.isinf(scores[line_id, allele_id]):\n",
    "                scores[line_id, allele_id] = 0.0\n",
    "            \n",
    "            for m in p3.finditer(chunk):\n",
    "                res_ind = res_start+m.start()\n",
    "                scores[line_id, allele_id] += points[res_ind]\n",
    "            counter += 1\n",
    "\n",
    "    return scores\n",
    "\n",
    "# blast crystal hla I seqs against hla data base\n",
    "# and create allele-pdb compatibility table\n",
    "def make_allele2pdb_table():\n",
    "    p4 = re.compile('.+([ABC]\\*[0-9]+:[0-9A-Z]+)')\n",
    "    alleles = []\n",
    "    file = open(\"../data/mhc_alleles/hla/hla1class.fasta\", 'r')\n",
    "    lines = file.readlines()\n",
    "    file.close()\n",
    "    for line in lines:\n",
    "        if line[0] == '>':\n",
    "            #alleles.append(p4.search(line).group(1))\n",
    "            alleles.append(line.split(';')[1])\n",
    "\n",
    "    parser = Bio.PDB.PDBParser(QUIET=True)\n",
    "    ppb = Bio.PDB.CaPPBuilder()\n",
    "\n",
    "    size = len(alleles)\n",
    "    num_alignments = size\n",
    "\n",
    "    scores = np.zeros((iclass.shape[0], size))\n",
    "    scores.fill(np.inf)\n",
    "\n",
    "    for i in range(iclass.shape[0]):\n",
    "        name = iclass.iloc[i, 0]\n",
    "        print(name)\n",
    "        chain = iclass.iloc[i, 3]\n",
    "        antichain = iclass.iloc[i, 7]\n",
    "        struct = parser.get_structure(id=name, file=pdb_path+'/'+name+'.pdb')[0]\n",
    "        seq = ''.join([str(x.get_sequence()) for x in ppb.build_peptides(struct[chain])])\n",
    "        query = open('../sandbox/query.fasta', 'w')\n",
    "        query.write('>%s\\n%s\\n' % (name, seq))\n",
    "        query.close()\n",
    "\n",
    "        command = [\"blastp\", \"-query\", \"../sandbox/query.fasta\",\n",
    "                   \"-subject\", \"../data/mhc_alleles/hla/hla1class.fasta\", \"-out\", \n",
    "                   \"../sandbox/out\", \"-outfmt\", str(3), \"-num_alignments\", str(num_alignments),\n",
    "                   \"-line_length\", str(1000), \"-ungapped\", \"-comp_based_stats\", \"F\"]\n",
    "        if subprocess.call(command) != 0:\n",
    "            print('Failed to blast')\n",
    "            exit()\n",
    "\n",
    "        points = np.array([0.0]*len(struct[chain]))\n",
    "        residues = list(struct[chain])\n",
    "        for j in range(len(residues)):\n",
    "            points[j] = get_points(residues[j], struct[antichain])\n",
    "\n",
    "        file = open(\"../sandbox/out\", 'r')\n",
    "        blasted = file.readlines()\n",
    "        file.close()\n",
    "\n",
    "        scores = score_alignments(scores, blasted, points, alleles, i, num_alignments)\n",
    "\n",
    "    alleles_vs_pdb = pd.DataFrame([alleles]+[list(x) for x in list(scores)]).T.drop_duplicates(0)\n",
    "    alleles_vs_pdb.index = alleles_vs_pdb[0]\n",
    "    alleles_vs_pdb = alleles_vs_pdb.iloc[:, 1:]\n",
    "    alleles_vs_pdb.columns = list(iclass.pdb)\n",
    "    alleles_vs_pdb = alleles_vs_pdb.replace(np.inf, np.nan)\n",
    "    alleles_vs_pdb.to_csv('../sandbox/alleles_vs_pdb', sep='\\t', float_format='%.2f')\n",
    "    \n",
    "#make_allele2pdb_table()\n",
    "\n",
    "alleles_vs_pdb = pd.read_csv('../sandbox/alleles_vs_pdb', sep='\\t', index_col=0, low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ligand_assays = pd.read_csv('../data/epitopes/mhc_ligand_full.csv', skiprows=1, low_memory=False)\n",
    "ligand_assays = ligand_assays[ligand_assays['MHC allele class'] == 'I']\n",
    "\n",
    "affinity = ligand_assays[['Description', 'Allele Name', 'Qualitative Measure', 'Quantitative measurement']]\n",
    "affinity['Length'] = [len(x) for x in affinity.Description]\n",
    "affinity = affinity[(affinity.Length < 12) & (affinity.Length > 7)]\n",
    "affinity = affinity[affinity['Allele Name'].str.contains('HLA-[ABC]\\*\\S+$', regex=True)]\n",
    "affinity['Allele Name'] = [x[4:] for x in affinity['Allele Name']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# compare peptides with bias to terminal acids\n",
    "def cmp_peptides(a, b):\n",
    "        assert(len(a) == len(b))\n",
    "        score = 0.0\n",
    "        length = len(a)\n",
    "        for i in range(length):\n",
    "            if a[i] != b[i]:\n",
    "                if i in [0,1,length-1]:\n",
    "                    score += 1.0\n",
    "                else:\n",
    "                    score += 0.5\n",
    "        return score\n",
    "\n",
    "# create peptide-pdb compatibility table\n",
    "def make_peptide2pdb_table():\n",
    "    antigen_list = list(set(affinity['Description']))\n",
    "    size = len(antigen_list)\n",
    "\n",
    "    peptide_scores = np.zeros((size, iclass.shape[0]))\n",
    "    peptide_scores.fill(np.inf)\n",
    "\n",
    "    for counter in range(size):\n",
    "        if counter % 1000 == 0:\n",
    "            print(counter)\n",
    "        seq = antigen_list[counter]\n",
    "        length = len(seq)\n",
    "\n",
    "        local = iclass[iclass.antigen_len == length]\n",
    "        for i in local.index:\n",
    "            peptide_scores[counter, i] = 0.0\n",
    "            crystal_seq = local.loc[i, 'antigen_seq']\n",
    "            peptide_scores[counter, i] = cmp_peptides(crystal_seq, seq)\n",
    "\n",
    "    peptides_vs_pdb = pd.DataFrame(peptide_scores, index=antigen_list, columns=list(iclass.pdb))\n",
    "    peptides_vs_pdb = peptides_vs_pdb.replace(np.inf, np.nan)\n",
    "    peptides_vs_pdb.to_csv('../sandbox/peptides_vs_pdb', sep='\\t')\n",
    "\n",
    "#make_peptide2pdb_table()\n",
    "\n",
    "peptides_vs_pdb = pd.read_csv('../sandbox/peptides_vs_pdb', sep='\\t', index_col=0, low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# s1 - mutated sequence\n",
    "# s2 - reference sequence\n",
    "def align(s1, s2):\n",
    "    matrix = matlist.blosum62\n",
    "    gap_open = -100.0\n",
    "    gap_extend = -0.5\n",
    "    alns = pairwise2.align.globalds(s1, s2, matrix, gap_open, gap_extend)\n",
    "    top_aln = alns[0]\n",
    "    \n",
    "    \n",
    "    print(top_aln[0])\n",
    "    print(top_aln[1])\n",
    "    \n",
    "    # check that there is no gaps inside the alignment\n",
    "    assert(re.search('[A-Z]-+[A-Z]', top_aln[0]) == None)\n",
    "    assert(re.search('[A-Z]-+[A-Z]', top_aln[1]) == None)\n",
    "    \n",
    "    mutations = []\n",
    "    gap = re.match('(-+)', top_aln[0])\n",
    "    shift = 0\n",
    "    if gap:\n",
    "        shift = gap.end(1) - gap.start(1)\n",
    "        \n",
    "    for i, a1, a2 in zip(list(range(-shift+1, len(top_aln[0])-shift+1)), top_aln[0], top_aln[1]):\n",
    "        if a1 != '-' and a2 != '-':\n",
    "            if a1 != a2:\n",
    "                mutations.append((str(i), a1, a2))\n",
    "    return mutations\n",
    "\n",
    "# mutate pdb structure in accordance to input sequences\n",
    "def mutate(pdb, newseq, newpep, table):\n",
    "    parser = Bio.PDB.PDBParser(QUIET=True)\n",
    "    ppb=Bio.PDB.PPBuilder()\n",
    "    \n",
    "    name = pdb; \n",
    "    amhc = table[table.pdb == pdb].iloc[0, 3]\n",
    "    pept = table[table.pdb == pdb].iloc[0, 7]\n",
    "    b2m = table[table.pdb == pdb].iloc[0, 5]\n",
    "    \n",
    "    struct = parser.get_structure(id=name, file=pdb_path+'/'+name+'.pdb')[0]\n",
    "    oldseq = ''.join([str(x.get_sequence()) for x in ppb.build_peptides(struct[amhc])])\n",
    "    oldpep = table[table.pdb == pdb].iloc[0, 8]\n",
    "    \n",
    "    assert(len(oldpep) == len(newpep))\n",
    "    mhc_mutations = align(oldseq, newseq)\n",
    "    pep_mutations = align(oldpep, newpep)\n",
    "\n",
    "    cmd.reinitialize()\n",
    "    cmd.load(pdb_path+'/'+name+'.pdb', 'obj')\n",
    "    \n",
    "    # remove alternative atom locations\n",
    "    cmd.remove(\"not alt ''+A\")\n",
    "    cmd.alter('all', \"alt=''\")\n",
    "    cmd.save('../sandbox/mutated/original.pdb', 'obj//%s// obj//%s// obj//%s//' % (amhc, pept, b2m))\n",
    "    \n",
    "    # mutate\n",
    "    cmd.wizard('mutagenesis')\n",
    "    cmd.refresh_wizard()\n",
    "\n",
    "    for i, old, new in mhc_mutations:\n",
    "        print(i, old, new)\n",
    "        cmd.get_wizard().do_select(amhc+'/'+str(i)+'/')\n",
    "        cmd.get_wizard().set_mode(seq3(new).upper())\n",
    "        cmd.get_wizard().apply()\n",
    "    print(\"\\n%s: %i MHC mutations\\n\" % (name, len(mhc_mutations)))\n",
    "        \n",
    "    for i, old, new in pep_mutations:\n",
    "        print(i, old, new)\n",
    "        cmd.get_wizard().do_select(pept+'/'+str(i)+'/')\n",
    "        cmd.get_wizard().set_mode(seq3(new).upper())\n",
    "        cmd.get_wizard().apply()\n",
    "    print(\"\\n%s: %i peptide mutations\\n\" % (name, len(pep_mutations)))\n",
    "        \n",
    "    cmd.set_wizard()\n",
    "    cmd.save('../sandbox/mutated/mutated.pdb', 'obj//%s// obj//%s// obj//%s//' % (amhc, pept, b2m))\n",
    "    #sleep(0.5)\n",
    "\n",
    "# wrapper for mutate(..)\n",
    "def find_and_mutate(pepseq, allele, alleledict, pep2pdb, mhc2pdb, pdb_table):\n",
    "    \n",
    "    chunk = mhc2pdb.loc[mhc2pdb.index.str.startswith(allele), :]\n",
    "    best_pdbs = (chunk.apply(min) + pep2pdb.loc[pepseq, :]/10).sort_values()\n",
    "    for i in xrange(len(best_pdbs)):\n",
    "        try:\n",
    "            best_pdb = best_pdbs.index[i]\n",
    "            allele_list = chunk.loc[:, best_pdb]\n",
    "            full_allele = allele_list[allele_list == min(allele_list)].index[0]\n",
    "            print(full_allele)\n",
    "            alleleseq = str(alleledict[full_allele].seq)\n",
    "            print('Compability score: %f' % best_pdbs[best_pdb])\n",
    "            mutate(best_pdb, alleleseq, pepseq, pdb_table)\n",
    "        except AssertionError:\n",
    "            print('Assertion error')\n",
    "            continue\n",
    "        else:\n",
    "            break\n",
    "    \n",
    "# testing\n",
    "alleledict = Bio.SeqIO.to_dict(Bio.SeqIO.parse(hla_fasta, 'fasta'))\n",
    "for key in alleledict.keys():\n",
    "    alleledict[key.split(';')[1]] = alleledict.pop(key)\n",
    "\n",
    "idx = 5000\n",
    "#for idx in xrange(1):\n",
    "pepseq = affinity.iloc[idx, 0]\n",
    "allele = affinity.iloc[idx, 1]\n",
    "find_and_mutate(pepseq, allele, alleledict, peptides_vs_pdb, alleles_vs_pdb, iclass)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Description                     YLVAKQATV\n",
       "Allele Name                       A*02:03\n",
       "Qualitative Measure         Positive-High\n",
       "Quantitative measurement             33.4\n",
       "Length                                  9\n",
       "Name: 9053, dtype: object"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "affinity.iloc[4000, :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "counter = 0\n",
    "for item in Bio.SeqIO.parse(\"../data/mhc_alleles/hla/hla1class.fasta\", 'fasta'):\n",
    "    if counter == 427:\n",
    "        print(item)\n",
    "        break\n",
    "    counter+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "mutate('4i48', alleledict['A*02:01'].seq, 'NLVPMVHTV', iclass)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'3mrb', '4i48'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pdb</th>\n",
       "      <th>mhc_class</th>\n",
       "      <th>host</th>\n",
       "      <th>a_chain_id</th>\n",
       "      <th>a_chain_allele</th>\n",
       "      <th>b_chain_id</th>\n",
       "      <th>b_chain_allele</th>\n",
       "      <th>antigen_id</th>\n",
       "      <th>antigen_seq</th>\n",
       "      <th>antigen_len</th>\n",
       "      <th>antigen_name</th>\n",
       "      <th>antigen_organism</th>\n",
       "      <th>comments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>3mrb</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>P</td>\n",
       "      <td>NLVPMVHTV</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     pdb mhc_class   host a_chain_id a_chain_allele b_chain_id b_chain_allele  \\\n",
       "53  3mrb         I  Human          A        A*02:01          B            B2M   \n",
       "\n",
       "   antigen_id antigen_seq  antigen_len antigen_name antigen_organism comments  \n",
       "53          P   NLVPMVHTV            9          NaN              NaN      NaN  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iclass[iclass.pdb == '3mrb']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "table=iclass\n",
    "pdb ='4hx1'\n",
    "parser = Bio.PDB.PDBParser(QUIET=True)\n",
    "ppb=Bio.PDB.PPBuilder()\n",
    "name = pdb; amhc = table[table.pdb == pdb].iloc[0, 3]; pept = table[table.pdb == pdb].iloc[0, 7]\n",
    "struct = parser.get_structure(id=name, file=pdb_path+'/'+name+'.pdb')[0]\n",
    "oldseq = str(ppb.build_peptides(struct[amhc])[0].get_sequence())\n",
    "oldpep = table[table.pdb == pdb].iloc[0, 8]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "str(ppb.build_peptides(struct[amhc])[0].get_sequence())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "affinity.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'B*44:03:22'"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alleledict.keys()[427]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "command = [\"blastp\", \"-query\", \"../testq\",\n",
    "                   \"-subject\", \"../tests\", \"-out\", \n",
    "                   \"../testout\", \"-outfmt\", str(3), \"-num_alignments\", str(20),\n",
    "                   \"-line_length\", str(1000), \"-ungapped\", \"-comp_based_stats\", \"F\"]\n",
    "if subprocess.call(command) != 0:\n",
    "    print('Failed to blast')\n",
    "    exit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda env:python2]",
   "language": "python",
   "name": "conda-env-python2-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
