{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "import Bio.PDB\n",
    "import Bio.SeqIO\n",
    "from Bio import pairwise2\n",
    "from Bio.SubsMat import MatrixInfo as matlist\n",
    "from Bio.SeqUtils import seq3\n",
    "\n",
    "FACTOR=10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# launch pymol\n",
    "\n",
    "import __main__\n",
    "__main__.pymol_argv = ['pymol','-Qc'] # Pymol: quiet and no GUI\n",
    "from time import sleep\n",
    "import pymol\n",
    "from pymol import cmd\n",
    "\n",
    "pymol.finish_launching()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "reload(sys)\n",
    "sys.setdefaultencoding('utf-8')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# paths\n",
    "\n",
    "hla_fasta = '../data/mhc_alleles/hla/hla_prot.fasta'\n",
    "pdb_path = '../data/structures'\n",
    "aligned_pdb_path = '../sandbox/aligned_pdb'\n",
    "corrected_pdb_path = '../sandbox/corrected_pdb'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# tables\n",
    "\n",
    "annotation = pd.read_csv('../sandbox/detailed_mhc', sep='\\t')\n",
    "alleles_vs_pdb = pd.read_csv('../sandbox/alleles_vs_pdb', sep='\\t', index_col=0, low_memory=False)\n",
    "peptides_vs_pdb = pd.read_csv('../sandbox/peptides_vs_pdb', sep='\\t', index_col=0, low_memory=False)\n",
    "ligand_assays = pd.read_csv('../data/epitopes/mhc_ligand_full.csv', skiprows=1, low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# table conversion\n",
    "\n",
    "ligand_assays = ligand_assays[ligand_assays['MHC allele class'] == 'I']\n",
    "affinity = ligand_assays[['Description', 'Allele Name', 'Qualitative Measure', 'Quantitative measurement']]\n",
    "affinity['Length'] = [len(x) for x in affinity.Description]\n",
    "affinity = affinity[(affinity.Length < 12) & (affinity.Length > 7)]\n",
    "affinity = affinity[affinity['Allele Name'].str.contains('HLA-[ABC]\\*\\S+$', regex=True)]\n",
    "affinity['Allele Name'] = [x[4:] for x in affinity['Allele Name']]\n",
    "\n",
    "iclass = annotation[annotation.mhc_class == 'I']\n",
    "iclass = iclass[iclass.a_chain_allele.str.contains('[ABC]\\*', regex=True)]\n",
    "iclass.a_chain_allele = [':'.join(x.split(':')[:2]) for x in iclass.a_chain_allele]\n",
    "iclass = iclass.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# place mhc to the specific position\n",
    "def localize_pdb(name, amhc, bmhc, pept, plen):\n",
    "    # define peptide terminal CAs\n",
    "    parser = Bio.PDB.PDBParser(QUIET=True)\n",
    "    ppb = Bio.PDB.CaPPBuilder() \n",
    "    struct = parser.get_structure(id=name, file=pdb_path+'/'+name+'.pdb')[0]\n",
    "    reslist = ppb.build_peptides(struct[pept])[0]\n",
    "    firstCA = reslist[0].get_id()[1]\n",
    "    lastCA = reslist[-1].get_id()[1]\n",
    "    \n",
    "    assert(lastCA-firstCA+1 == plen)\n",
    "    \n",
    "    # translate complex by -startCA\n",
    "    cmd.select('sele', '%s//%s/%i/CA %s//%s/%i/CA' % (name, pept, firstCA, name, pept, lastCA))\n",
    "    start, end = cmd.get_coords('sele', 1)\n",
    "    cmd.translate(list(-start), name)\n",
    "    \n",
    "    # rotate complex around (0, 0, 0), so that endCA.y = 0\n",
    "    start, end = cmd.get_coords('sele', 1)\n",
    "    angle = np.arctan(end[1]/end[0])\n",
    "    if end[0] < 0.0:\n",
    "        angle = np.pi + angle\n",
    "    angle = -angle*180.0/np.pi\n",
    "    cmd.rotate('z', angle, name, origin=[0.0, 0.0, 0.0])\n",
    "    \n",
    "    # superpose peptide line on x axis\n",
    "    start, end = cmd.get_coords('sele', 1)\n",
    "    angle = np.arctan(end[2]/end[0])\n",
    "    if end[0] < 0.0:\n",
    "        angle = np.pi + angle\n",
    "    angle = angle*180.0/np.pi\n",
    "    print(cmd.rotate('y', angle, name, origin=[0.0, 0.0, 0.0]))\n",
    "    \n",
    "    # shift along x axis\n",
    "    start, end = cmd.get_coords('sele', 1)\n",
    "    print(cmd.translate([-(start[0]+end[0])/2.0, 0.0, 0.0], name))\n",
    "    \n",
    "    # superpose com on the xz plane\n",
    "    com = cmd.centerofmass('%s//%s// %s//%s//' % (name, amhc, name, bmhc))\n",
    "    angle = np.arctan(com[2]/com[1])\n",
    "    if com[1] < 0.0:\n",
    "        angle = np.pi + angle\n",
    "    angle = -(angle)*180.0/np.pi\n",
    "    print(cmd.rotate('x', angle, name, origin=[0.0, 0.0, 0.0]))\n",
    "    \n",
    "    # verify the result\n",
    "    com = cmd.centerofmass('%s//%s// %s//%s//' % (name, amhc, name, bmhc))\n",
    "    start, end = cmd.get_coords('sele', 1)\n",
    "    assert(com[2] < 0.001)\n",
    "    assert(start[2] < 0.001 and start[1] < 0.001)\n",
    "    assert(end[2] < 0.001 and end[1] < 0.001)\n",
    "    \n",
    "    #cmd.delete('!(%s//%s// %s//%s// %s//%s//)' % (name, mhca, name, mhcb, name, pept))\n",
    "    \n",
    "    #cmd.save('../sandbox/positioned_pdb/%s_posed.pdb' % name, \n",
    "    #         '%s//%s// %s//%s// %s//%s//' % (name, mhca, name, mhcb, name, pept))\n",
    "    \n",
    "    sleep(0.5)\n",
    "\n",
    "# name should be preloaded to pymol with cmd.load()    \n",
    "def localize_pdb_wrapper(name, table):\n",
    "    line = table[table.pdb == name].iloc[0, :]\n",
    "    mhca = line['a_chain_id']\n",
    "    mhcb = line['b_chain_id']\n",
    "    pept = line['antigen_id']\n",
    "    plen = int(line['antigen_len'])\n",
    "\n",
    "    localize_pdb(name, mhca, mhcb, pept, plen)\n",
    "    \n",
    "# name should be preloaded to pymol with cmd.load()    \n",
    "def clean(name, table):\n",
    "    line = table[table.pdb == name].iloc[0, :]\n",
    "    mhca = line['a_chain_id']\n",
    "    mhcb = line['b_chain_id']\n",
    "    pept = line['antigen_id']\n",
    "    \n",
    "    cmd.delete('!(%s//%s// %s//%s// %s//%s//)' % (name, mhca, name, mhcb, name, pept))\n",
    "    \n",
    "# reference and mobile should be preloaded to pymol with cmd.load()\n",
    "def align_mob2ref(mob_name, ref_name, table):\n",
    "    line = table[table.pdb == ref_name].iloc[0, :]\n",
    "    ref_mhca = line['a_chain_id']\n",
    "    ref_mhcb = line['b_chain_id']\n",
    "    ref_pept = line['antigen_id']\n",
    "    \n",
    "    line = table[table.pdb == mob_name].iloc[0, :]\n",
    "    mob_mhca = line['a_chain_id']\n",
    "    mob_mhcb = line['b_chain_id']\n",
    "    mob_pept = line['antigen_id']\n",
    "\n",
    "    cmd.align('%s//%s// %s//%s//' % (mob_name, mob_mhca, mob_name, mob_pept), \n",
    "              '%s//%s// %s//%s//' % (ref_name, ref_mhca, ref_name, ref_pept))\n",
    "    \n",
    "# name should be preloaded to pymol with cmd.load()\n",
    "def save2pdb(name, table, path):\n",
    "    line = table[table.pdb == name].iloc[0, :]\n",
    "    mhca = line['a_chain_id']\n",
    "    mhcb = line['b_chain_id']\n",
    "    pept = line['antigen_id']\n",
    "    \n",
    "    cmd.save(path, '%s//%s// %s//%s// %s//%s//' % (name, mhca, name, mhcb, name, pept)) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def align_all_pdbs(table):\n",
    "    ref=table.loc[0, 'pdb']\n",
    "\n",
    "    cmd.reinitialize()\n",
    "    cmd.load(pdb_path+'/'+ref+'.pdb', ref)\n",
    "    clean(ref, table)\n",
    "    localize_pdb_wrapper(ref, table)\n",
    "\n",
    "    for i in range(1, table.shape[0]):\n",
    "        name=table.loc[i, 'pdb']\n",
    "        cmd.load(pdb_path+'/'+name+'.pdb', name)\n",
    "        clean(name, table)\n",
    "        align_mob2ref(name, ref, table)\n",
    "        save2pdb(name, table, '%s/%s_aligned.pdb' % (aligned_pdb_path, name))\n",
    "        cmd.delete(name)\n",
    "\n",
    "    save2pdb(ref, table, '%s/%s_aligned.pdb' % (aligned_pdb_path, ref))\n",
    "    cmd.delete(ref)\n",
    "        \n",
    "#align_all_pdbs(iclass)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pdb</th>\n",
       "      <th>mhc_class</th>\n",
       "      <th>host</th>\n",
       "      <th>a_chain_id</th>\n",
       "      <th>a_chain_allele</th>\n",
       "      <th>b_chain_id</th>\n",
       "      <th>b_chain_allele</th>\n",
       "      <th>antigen_id</th>\n",
       "      <th>antigen_seq</th>\n",
       "      <th>antigen_len</th>\n",
       "      <th>antigen_name</th>\n",
       "      <th>antigen_organism</th>\n",
       "      <th>comments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3i6l</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*24:02</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>QFKDNVILL</td>\n",
       "      <td>9</td>\n",
       "      <td>Nucleoprotein</td>\n",
       "      <td>SARS coronavirus Tor2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5swq</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>CVNGSCFTV</td>\n",
       "      <td>9</td>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>Influenza A virus</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3x14</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>B*08:02</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>FLRGRAYGL</td>\n",
       "      <td>9</td>\n",
       "      <td>nuclear antigen EBNA-3</td>\n",
       "      <td>Human herpesvirus 4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4jqx</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>B*44:03</td>\n",
       "      <td>C</td>\n",
       "      <td>B2M</td>\n",
       "      <td>B</td>\n",
       "      <td>EECDSELEIKRY</td>\n",
       "      <td>12</td>\n",
       "      <td>Trans-activator protein BZLF1</td>\n",
       "      <td>Human herpesvirus 4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5c0d</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>AQWGPDPAAA</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    pdb mhc_class   host a_chain_id a_chain_allele b_chain_id b_chain_allele  \\\n",
       "0  3i6l         I  Human          D        A*24:02          E            B2M   \n",
       "1  5swq         I  Human          A        A*02:01          B            B2M   \n",
       "2  3x14         I  Human          A        B*08:02          B            B2M   \n",
       "3  4jqx         I  Human          A        B*44:03          C            B2M   \n",
       "4  5c0d         I  Human          A        A*02:01          B            B2M   \n",
       "\n",
       "  antigen_id   antigen_seq  antigen_len                   antigen_name  \\\n",
       "0          F     QFKDNVILL            9                  Nucleoprotein   \n",
       "1          C     CVNGSCFTV            9                  neuraminidase   \n",
       "2          C     FLRGRAYGL            9         nuclear antigen EBNA-3   \n",
       "3          B  EECDSELEIKRY           12  Trans-activator protein BZLF1   \n",
       "4          C    AQWGPDPAAA           10                            NaN   \n",
       "\n",
       "        antigen_organism comments  \n",
       "0  SARS coronavirus Tor2      NaN  \n",
       "1      Influenza A virus      NaN  \n",
       "2    Human herpesvirus 4      NaN  \n",
       "3    Human herpesvirus 4      NaN  \n",
       "4                    NaN      NaN  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iclass.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def set_values(matrix, atom_desc, left_bottom, right_up, bin_size):\n",
    "    aas = ['GLY', 'ALA', 'ILE', 'LEU', \n",
    "           'PRO', 'SER', 'THR', 'CYS', \n",
    "           'MET', 'ASP', 'ASN', 'GLU', \n",
    "           'GLN', 'LYS', 'ARG', 'HIS', \n",
    "           'PHE', 'TYR', 'TRP']\n",
    "    \n",
    "    x = atom_desc[5]\n",
    "    y = atom_desc[6]\n",
    "    z = atom_desc[7]\n",
    "    \n",
    "    # check if the atom is inside of the box\n",
    "    if (x-left_bottom[0] < 0.0 or y-left_bottom[1] < 0.0 or z-left_bottom[2] < 0.0):\n",
    "        return 0\n",
    "    \n",
    "    if (x-right_up[0] >= 0.0 or y-right_up[1] >= 0.0 or z-right_up[2] >= 0.0):\n",
    "        return 0\n",
    "    \n",
    "    # compute atom index\n",
    "    x_bin = np.floor((x-left_bottom[0])/bin_size)\n",
    "    y_bin = np.floor((y-left_bottom[1])/bin_size)\n",
    "    z_bin = np.floor((z-left_bottom[2])/bin_size)\n",
    "    \n",
    "    #if y_bin == 20:\n",
    "    #        print(y-left_bottom[1])\n",
    "    \n",
    "    # increase values in matrix corresponding to the atom\n",
    "    if atom_desc[0] != 'ATOM':\n",
    "        matrix['X'][x_bin, y_bin, z_bin] += 1.0\n",
    "        matrix['XXX'][x_bin, y_bin, z_bin] += 1.0\n",
    "        return 1\n",
    "    \n",
    "    if atom_desc[1] == 'CA':\n",
    "        matrix['CA'][x_bin, y_bin, z_bin] += 1.0\n",
    "        \n",
    "    if atom_desc[2] in ['C', 'N', 'O', 'S']:\n",
    "        matrix[atom_desc[2]][x_bin, y_bin, z_bin] += 1.0\n",
    "    else:\n",
    "        matrix['X'][x_bin, y_bin, z_bin] += 1.0\n",
    "    \n",
    "    if atom_desc[3] in aas:\n",
    "        matrix[atom_desc[3]][x_bin, y_bin, z_bin] += 1.0\n",
    "    else:\n",
    "        matrix['XXX'][x_bin, y_bin, z_bin] += 1.0\n",
    "        \n",
    "    if atom_desc[4] == 'H':\n",
    "        matrix['HELIX'][x_bin, y_bin, z_bin] += 1.0\n",
    "        \n",
    "    if atom_desc[4] == 'S':\n",
    "        matrix['STRAND'][x_bin, y_bin, z_bin] += 1.0\n",
    "        \n",
    "    return 2\n",
    "\n",
    "# convert complex to 4d matrix\n",
    "# name should be preloaded to pymol with cmd.load()\n",
    "def mhc2matrix(name, left_bottom=(-30.0, -10.0, -10.0), right_up=(30.0, 10.0, 10.0), bin_size=1.0):\n",
    "    x_bins = int((right_up[0]-left_bottom[0])/bin_size)\n",
    "    y_bins = int((right_up[1]-left_bottom[1])/bin_size)\n",
    "    z_bins = int((right_up[2]-left_bottom[2])/bin_size)\n",
    "    \n",
    "    layer_names = ['CA', 'C', 'N', 'O', 'S', 'X', \n",
    "                   'GLY', 'ALA', 'ILE', 'LEU', \n",
    "                   'PRO', 'SER', 'THR', 'CYS', \n",
    "                   'MET', 'ASP', 'ASN', 'GLU', \n",
    "                   'GLN', 'LYS', 'ARG', 'HIS', \n",
    "                   'PHE', 'TYR', 'TRP', 'XXX', \n",
    "                   'HELIX', 'STRAND']\n",
    "    \n",
    "    filter_num = len(layer_names)\n",
    "    m = np.zeros((x_bins, y_bins, z_bins), dtype=zip(layer_names, [np.float32]*len(layer_names)))\n",
    "    \n",
    "    atoms = []\n",
    "    myspace = {'atoms': atoms}\n",
    "    cmd.iterate_state(0, name, 'atoms.append([type, name, elem, resn, ss, x, y, z])', space=myspace)\n",
    "    #print(atoms)\n",
    "    \n",
    "    for atom_desc in atoms: #cmd.iterate_state(0, name+'//C//', 'type, name, elem, resn, ss, x, y, z'):\n",
    "        set_values(m, atom_desc, left_bottom, right_up, bin_size)\n",
    "        \n",
    "    return np.array([m[name] for name in layer_names])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# select terminal peptide CA and all atoms in range of 15 A from the peptide\n",
    "def select_atoms(name, pept, plen):\n",
    "    parser = Bio.PDB.PDBParser(QUIET=True)\n",
    "    ppb = Bio.PDB.CaPPBuilder() \n",
    "    struct = parser.get_structure(id=name, file=corrected_pdb_path+'/'+name+'.pdb')[0]\n",
    "    reslist = ppb.build_peptides(struct[pept])[0]\n",
    "    firstCA = reslist[0].get_id()[1]\n",
    "    lastCA = reslist[-1].get_id()[1]\n",
    "    \n",
    "    assert(lastCA-firstCA+1 == plen)\n",
    "    \n",
    "    cmd.select('fixed', '%s//%s/%i/CA %s//%s/%i/CA byres %s be. 15 of %s//%s//' % (name, pept, firstCA, \n",
    "                                                                                   name, pept, lastCA, \n",
    "                                                                                   name, name, pept))\n",
    "    idx = []\n",
    "    myspace = {'idx': idx}\n",
    "    cmd.iterate_state(0, name, 'idx.append(ID)', space=myspace)  \n",
    "    \n",
    "    return idx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_best_pdbs(allele, pepseq, mhc2pdb, pep2pdb):\n",
    "    chunk = mhc2pdb.loc[mhc2pdb.index.str.startswith(allele), :]\n",
    "    best_pdbs = (chunk.apply(min) + pep2pdb.loc[pepseq, :]/10).sort_values()\n",
    "    return best_pdbs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def ungapped_align(a, b, matrix, peptide=False):\n",
    "    score = -np.inf\n",
    "    pos = []\n",
    "    \n",
    "    if not peptide:\n",
    "        for i in range(len(a) + len(b)):\n",
    "            a_beg, a_end = (max(0, i-len(b)), min(i, len(a)))\n",
    "            b_beg, b_end = (max(0, len(b)-i), min(len(b), len(b)+len(a)-i))\n",
    "\n",
    "            loc_score = sum([matrix[(k, t)] for k, t in zip(a[a_beg:a_end], b[b_beg:b_end])])\n",
    "            if loc_score >= score:\n",
    "                    score = loc_score\n",
    "                    pos = [a_beg, b_beg]\n",
    "    else:\n",
    "        pos = [0, 0]\n",
    "    \n",
    "    a_aligned = a\n",
    "    b_aligned = b\n",
    "    \n",
    "    if (pos[0] > pos[1]):\n",
    "        b_aligned = '-'*(pos[0]-pos[1])+b_aligned\n",
    "    else: \n",
    "        a_aligned = '-'*(pos[1]-pos[0])+a_aligned\n",
    "        \n",
    "    if (len(a_aligned) > len(b_aligned)):\n",
    "        b_aligned += '-'*(len(a_aligned) - len(b_aligned))\n",
    "    else:\n",
    "        a_aligned += '-'*(len(b_aligned) - len(a_aligned))\n",
    "    \n",
    "    return a_aligned, b_aligned\n",
    "\n",
    "# s1 - mutated sequence\n",
    "# s2 - reference sequence\n",
    "def align(s1, s2, peptide=False):\n",
    "    matrix = {}\n",
    "    for key, val in matlist.blosum62.iteritems(): # do it outside the function\n",
    "        matrix.update({(key[1], key[0]): val})\n",
    "        matrix.update({(key[0], key[1]): val})\n",
    "    \n",
    "    top_aln = ungapped_align(s1, s2, matrix, peptide)\n",
    "\n",
    "    print(\"\\tOld: \"+top_aln[0])\n",
    "    print(\"\\tNew: \"+top_aln[1])\n",
    "    \n",
    "    # check that there is no gaps inside the alignment\n",
    "    assert(re.search('[A-Z]-+[A-Z]', top_aln[0]) == None)\n",
    "    assert(re.search('[A-Z]-+[A-Z]', top_aln[1]) == None)\n",
    "    \n",
    "    mutations = []\n",
    "    gap = re.match('(-+)', top_aln[0])\n",
    "    shift = 0\n",
    "    if gap:\n",
    "        shift = gap.end(1) - gap.start(1)\n",
    "        \n",
    "    for i, a1, a2 in zip(list(range(-shift, len(top_aln[0])-shift)), top_aln[0], top_aln[1]):\n",
    "        if a1 != '-' and a2 != '-':\n",
    "            if a1 != a2:\n",
    "                mutations.append((i, a1, a2))\n",
    "    return mutations\n",
    "\n",
    "'''def align(s1, s2):\n",
    "    matrix = matlist.blosum62\n",
    "    gap_open = -100.0\n",
    "    gap_extend = -1.0\n",
    "    alns = pairwise2.align.globalds(s1, s2, matrix, gap_open, gap_extend)\n",
    "    top_aln = alns[0]\n",
    "    \n",
    "    \n",
    "    print(top_aln[0])\n",
    "    print(top_aln[1])\n",
    "    \n",
    "    # check that there is no gaps inside the alignment\n",
    "    assert(re.search('[A-Z]-+[A-Z]', top_aln[0]) == None)\n",
    "    assert(re.search('[A-Z]-+[A-Z]', top_aln[1]) == None)\n",
    "    \n",
    "    mutations = []\n",
    "    gap = re.match('(-+)', top_aln[0])\n",
    "    shift = 0\n",
    "    if gap:\n",
    "        shift = gap.end(1) - gap.start(1)\n",
    "        \n",
    "    for i, a1, a2 in zip(list(range(-shift, len(top_aln[0])-shift)), top_aln[0], top_aln[1]):\n",
    "        if a1 != '-' and a2 != '-':\n",
    "            if a1 != a2:\n",
    "                mutations.append((i, a1, a2))\n",
    "    return mutations'''\n",
    "\n",
    "# mutate pdb structure in accordance to input sequences\n",
    "# here we assume that all chains are continuous: CHECK THIS\n",
    "def mutate(pdb, newseq, newpep, table):\n",
    "    parser = Bio.PDB.PDBParser(QUIET=True)\n",
    "    ppb=Bio.PDB.PPBuilder()\n",
    "    \n",
    "    name = pdb; \n",
    "    amhc = table[table.pdb == pdb].iloc[0, 3]\n",
    "    pept = table[table.pdb == pdb].iloc[0, 7]\n",
    "    b2m = table[table.pdb == pdb].iloc[0, 5]\n",
    "    \n",
    "    struct = parser.get_structure(id=name, file=pdb_path+'/'+name+'.pdb')[0]\n",
    "    oldseq = ''.join([str(x.get_sequence()) for x in ppb.build_peptides(struct[amhc])])\n",
    "    oldpep = table[table.pdb == pdb].iloc[0, 8]\n",
    "    \n",
    "    assert(len(oldpep) == len(newpep))\n",
    "    print(\"\\n======= MHC business =======\\n\")\n",
    "    mhc_mutations = align(oldseq, newseq)\n",
    "    print(\"\\nMHC mutations: %i\\n\" % len(mhc_mutations))\n",
    "    \n",
    "    #assert(len(mhc_mutations) < 30)\n",
    "    \n",
    "    print(\"\\n===== Peptide business =====\\n\")\n",
    "    pep_mutations = align(oldpep, newpep, True)\n",
    "    print(\"\\nPeptide mutations: %i\\n\\n\" % len(pep_mutations))\n",
    "    \n",
    "    print(\"Mutagenesis: \\n\")\n",
    "    # remove alternative atom locations\n",
    "    cmd.remove(\"not alt ''+A\")\n",
    "    cmd.alter('all', \"alt=''\")\n",
    "    \n",
    "    # mutate\n",
    "    cmd.wizard('mutagenesis')\n",
    "    cmd.refresh_wizard()\n",
    "\n",
    "    first_aa_id = ppb.build_peptides(struct[amhc])[0][0].get_id()[1]\n",
    "    for i, old, new in mhc_mutations:\n",
    "        print(i, old, new)\n",
    "        cmd.get_wizard().do_select(amhc+'/'+str(i+first_aa_id)+'/')\n",
    "        cmd.get_wizard().set_mode(seq3(new).upper())\n",
    "        cmd.get_wizard().apply()\n",
    "        \n",
    "    first_aa_id = ppb.build_peptides(struct[pept])[0][0].get_id()[1]\n",
    "    for i, old, new in pep_mutations:\n",
    "        print(i, old, new)\n",
    "        cmd.get_wizard().do_select(pept+'/'+str(i+first_aa_id)+'/')\n",
    "        cmd.get_wizard().set_mode(seq3(new).upper())\n",
    "        cmd.get_wizard().apply()\n",
    "        \n",
    "    cmd.set_wizard()\n",
    "    \n",
    "    #sleep(0.5)\n",
    "    return (len(mhc_mutations), len(pep_mutations)) # for tests TEMPORARY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#allele = iclass[iclass.pdb == '5brz']['a_chain_allele'].iloc[0]\n",
    "#pepseq = iclass[iclass.pdb == '5brz']['antigen_seq'].iloc[0]\n",
    "#chunk = alleles_vs_pdb.loc[alleles_vs_pdb.index.str.startswith(allele), :]\n",
    "#best_pdbs = (chunk.apply(min) + peptides_vs_pdb.loc[pepseq, :]/FACTOR).sort_values()\n",
    "#chunk['3bo8']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "alleledict = Bio.SeqIO.to_dict(Bio.SeqIO.parse(hla_fasta, 'fasta'))\n",
    "for key in alleledict.keys():\n",
    "    alleledict[key.split(';')[1]] = alleledict.pop(key)\n",
    "\n",
    "# wrapper for mutate(..)\n",
    "def find_and_mutate(pepseq, allele, alleledict, pep2pdb, mhc2pdb, pdb_table, first_best=10, ref=None):\n",
    "    if ref:\n",
    "        print(\"\\nChanging to \"+ref+\"\\n\")\n",
    "        allele = pdb_table[pdb_table.pdb == ref]['a_chain_allele'].iloc[0]\n",
    "        pepseq = pdb_table[pdb_table.pdb == ref]['antigen_seq'].iloc[0]\n",
    "    \n",
    "    chunk = mhc2pdb.loc[mhc2pdb.index.str.startswith(allele), :]\n",
    "    best_pdbs = (chunk.apply(min) + pep2pdb.loc[pepseq, :]/FACTOR).sort_values()\n",
    "    \n",
    "    ############# for tests TEMPORARY\n",
    "    report = open('../sandbox/mutated/report', 'a')\n",
    "    report.write('name\\tpdb\\ta_chain\\told_allele\\tnew_allele\\tmhc_mutations\\tpeptide_chain\\told_peptide\\tnew_peptide\\tpeptide_mutations\\tcompatibility_score\\trmsd\\n')\n",
    "    #############\n",
    "    \n",
    "    counter = 0\n",
    "    for i in xrange(len(best_pdbs)):\n",
    "        try:\n",
    "            print(\"\\n=============================================\")    \n",
    "            \n",
    "            best_pdb = best_pdbs.index[i]\n",
    "            comp_score = best_pdbs[best_pdb]\n",
    "            \n",
    "            if np.isnan(comp_score):\n",
    "                print(\"No compatible structures found\")\n",
    "                break\n",
    "            \n",
    "            print(\"Operating on \"+best_pdb+\"\\n\")    \n",
    "                \n",
    "            allele_list = chunk.loc[:, best_pdb]\n",
    "            full_allele = allele_list[allele_list == min(allele_list)].index[0]\n",
    "            print(\"New allele: \"+full_allele)\n",
    "            \n",
    "            alleleseq = str(alleledict[full_allele].seq)\n",
    "            print('Compability score: %f' % comp_score)\n",
    "            \n",
    "            ####################### for tests TEMPORARY\n",
    "            rmsd = peptides_rmsd(corrected_pdb_path+'/'+ref+'.pdb', \n",
    "                                 pdb_table[pdb_table.pdb == ref]['antigen_id'].iloc[0], \n",
    "                                 corrected_pdb_path+'/'+best_pdb+'.pdb', \n",
    "                                 pdb_table[pdb_table.pdb == best_pdb]['antigen_id'].iloc[0])\n",
    "            \n",
    "            print(\"Peptide RMSD: \"+str(rmsd))\n",
    "            \n",
    "            #if (rmsd <= 1.3 or rmsd > 1.6 or allele == pdb_table[pdb_table.pdb == best_pdb]['a_chain_allele'].iloc[0]):\n",
    "            #    continue\n",
    "            \n",
    "            cmd.save('../sandbox/mutated/ref_%s.pdb' % ref, 'obj1')\n",
    "            #######################\n",
    "            \n",
    "            cmd.reinitialize()\n",
    "            cmd.load(corrected_pdb_path+'/'+best_pdb+'.pdb', best_pdb)\n",
    "            #cmd.save('../sandbox/mutated/original%i.pdb' % counter, best_pdb)\n",
    "            \n",
    "            mhc_mut_num, pep_mut_num = mutate(best_pdb, alleleseq, pepseq, pdb_table)\n",
    "            line = pdb_table[pdb_table.pdb == best_pdb].iloc[0, :]\n",
    "            \n",
    "            ######################## for tests TEMPORARY\n",
    "            report.write('%s\\t%s\\t%s\\t%s\\t%s\\t%i\\t%s\\t%s\\t%s\\t%i\\t%.1f\\t%.2f\\n' % \n",
    "                         ('mutated%i.pdb' % counter, best_pdb, line.a_chain_id, line.a_chain_allele, allele, mhc_mut_num,\n",
    "                          line.antigen_id, line.antigen_seq, pepseq, pep_mut_num, comp_score, rmsd))\n",
    "            ########################\n",
    "            \n",
    "            cmd.save('../sandbox/mutated/mutated%i.pdb' % counter, best_pdb)\n",
    "            \n",
    "            # write fixed atoms for minimization protocol\n",
    "            idx = select_atoms(best_pdb, line.antigen_id, line.antigen_len)\n",
    "            with open('../sandbox/mutated/fixed_atoms%i' % counter, 'w') as f:\n",
    "                for ind in idx:\n",
    "                    f.write(\"%i\\n\" % ind)\n",
    "            \n",
    "            cmd.delete(best_pdb)\n",
    "            \n",
    "            # HERE WE SHOULD HAVE MINIMIZER INVOKED\n",
    "            \n",
    "        except AssertionError:\n",
    "            print('Assertion error: %s, %s, %s' % (pepseq, allele, best_pdb))\n",
    "            continue\n",
    "    \n",
    "        except pymol.CmdException:\n",
    "            print('Pymol error: %s, %s, %s' % (pepseq, allele, best_pdb))\n",
    "            continue\n",
    "        else:\n",
    "            if counter == first_best-1:\n",
    "                report.close()\n",
    "                break\n",
    "                \n",
    "            else:\n",
    "                counter += 1\n",
    "    \n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "find_and_mutate(None, None, alleledict, peptides_vs_pdb, alleles_vs_pdb, iclass, first_best=10, ref='5brz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# allele dictionary\n",
    "alleledict = Bio.SeqIO.to_dict(Bio.SeqIO.parse(hla_fasta, 'fasta'))\n",
    "for key in alleledict.keys():\n",
    "    alleledict[key.split(';')[1]] = alleledict.pop(key)\n",
    "    \n",
    "idx = 100000\n",
    "pepseq = affinity.iloc[idx, 0]\n",
    "allele = affinity.iloc[idx, 1]\n",
    "#ms = find_and_mutate(pepseq, allele, alleledict, peptides_vs_pdb, alleles_vs_pdb, iclass)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pdb</th>\n",
       "      <th>mhc_class</th>\n",
       "      <th>host</th>\n",
       "      <th>a_chain_id</th>\n",
       "      <th>a_chain_allele</th>\n",
       "      <th>b_chain_id</th>\n",
       "      <th>b_chain_allele</th>\n",
       "      <th>antigen_id</th>\n",
       "      <th>antigen_seq</th>\n",
       "      <th>antigen_len</th>\n",
       "      <th>antigen_name</th>\n",
       "      <th>antigen_organism</th>\n",
       "      <th>comments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>1q94</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*11:01</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>AIFQSSMTK</td>\n",
       "      <td>9</td>\n",
       "      <td>Pol polyprotein</td>\n",
       "      <td>Human immunodeficiency virus 1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>4mj6</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*11:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>TMVMELIRMIK</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>4n8v</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*11:01</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>MLIYSMWGK</td>\n",
       "      <td>9</td>\n",
       "      <td>Virion envelope protein A14</td>\n",
       "      <td>Vaccinia virus Copenhagen</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>4mj5</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*11:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>TIAMELIRMIK</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>1qvo</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*11:01</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>QVPLRPMTYK</td>\n",
       "      <td>10</td>\n",
       "      <td>Protein Nef</td>\n",
       "      <td>Human immunodeficiency virus type 1 (JRCSF ISO...</td>\n",
       "      <td>Imported from the TopBank database.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>2hn7</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*11:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>AIMPARFYPK</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>This peptide is an analog of the naturally occ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>1x7q</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*11:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>KTFPPTEPK</td>\n",
       "      <td>9</td>\n",
       "      <td>Nucleoprotein</td>\n",
       "      <td>SARS coronavirus Tor2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      pdb mhc_class   host a_chain_id a_chain_allele b_chain_id  \\\n",
       "42   1q94         I  Human          D        A*11:01          E   \n",
       "44   4mj6         I  Human          A        A*11:01          B   \n",
       "60   4n8v         I  Human          D        A*11:01          E   \n",
       "97   4mj5         I  Human          A        A*11:01          B   \n",
       "162  1qvo         I  Human          D        A*11:01          E   \n",
       "205  2hn7         I  Human          A        A*11:01          B   \n",
       "247  1x7q         I  Human          A        A*11:01          B   \n",
       "\n",
       "    b_chain_allele antigen_id  antigen_seq  antigen_len  \\\n",
       "42             B2M          F    AIFQSSMTK            9   \n",
       "44             B2M          C  TMVMELIRMIK           11   \n",
       "60             B2M          F    MLIYSMWGK            9   \n",
       "97             B2M          C  TIAMELIRMIK           11   \n",
       "162            B2M          F   QVPLRPMTYK           10   \n",
       "205            B2M          C   AIMPARFYPK           10   \n",
       "247            B2M          C    KTFPPTEPK            9   \n",
       "\n",
       "                    antigen_name  \\\n",
       "42               Pol polyprotein   \n",
       "44                           NaN   \n",
       "60   Virion envelope protein A14   \n",
       "97                           NaN   \n",
       "162                  Protein Nef   \n",
       "205                          NaN   \n",
       "247                Nucleoprotein   \n",
       "\n",
       "                                      antigen_organism  \\\n",
       "42                      Human immunodeficiency virus 1   \n",
       "44                                                 NaN   \n",
       "60                           Vaccinia virus Copenhagen   \n",
       "97                                                 NaN   \n",
       "162  Human immunodeficiency virus type 1 (JRCSF ISO...   \n",
       "205                                                NaN   \n",
       "247                              SARS coronavirus Tor2   \n",
       "\n",
       "                                              comments  \n",
       "42                                                 NaN  \n",
       "44                                                 NaN  \n",
       "60                                                 NaN  \n",
       "97                                                 NaN  \n",
       "162                Imported from the TopBank database.  \n",
       "205  This peptide is an analog of the naturally occ...  \n",
       "247                                                NaN  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iclass[iclass.a_chain_allele == 'A*11:01']\n",
    "#iclass[iclass.pdb == '1a1m']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 438,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pdb</th>\n",
       "      <th>mhc_class</th>\n",
       "      <th>host</th>\n",
       "      <th>a_chain_id</th>\n",
       "      <th>b_chain_id</th>\n",
       "      <th>b_chain_allele</th>\n",
       "      <th>antigen_id</th>\n",
       "      <th>antigen_seq</th>\n",
       "      <th>antigen_len</th>\n",
       "      <th>antigen_name</th>\n",
       "      <th>antigen_organism</th>\n",
       "      <th>comments</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>a_chain_allele</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A*01:01</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A*02:01</th>\n",
       "      <td>186</td>\n",
       "      <td>186</td>\n",
       "      <td>186</td>\n",
       "      <td>186</td>\n",
       "      <td>186</td>\n",
       "      <td>186</td>\n",
       "      <td>186</td>\n",
       "      <td>186</td>\n",
       "      <td>186</td>\n",
       "      <td>101</td>\n",
       "      <td>101</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A*02:03</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A*02:06</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A*02:07</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A*02:30</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A*02:562</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A*02:66</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A*03:01</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A*11:01</th>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A*24:02</th>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A*68:01</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A*68:02</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A*68:12</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*07:02</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*08:01</th>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*08:02</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*08:136</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*14:02</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*15:01</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*18:01</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*27:04</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*27:05</th>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*27:06</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*27:09</th>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*27:13</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*35:01</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*35:08</th>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*35:42</th>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*39:01</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*41:43</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*42:01</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*42:02</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*44:02</th>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*44:03</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*44:05</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*46:01</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*51:193</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*53:01</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*57:01</th>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*57:03</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*57:06</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*58:01</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B*81:01</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C*03:04</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C*04:09N</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C*08:22</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                pdb  mhc_class  host  a_chain_id  b_chain_id  b_chain_allele  \\\n",
       "a_chain_allele                                                                 \n",
       "A*01:01           6          6     6           6           6               6   \n",
       "A*02:01         186        186   186         186         186             186   \n",
       "A*02:03           1          1     1           1           1               1   \n",
       "A*02:06           1          1     1           1           1               1   \n",
       "A*02:07           1          1     1           1           1               1   \n",
       "A*02:30           1          1     1           1           1               1   \n",
       "A*02:562          1          1     1           1           1               1   \n",
       "A*02:66           1          1     1           1           1               1   \n",
       "A*03:01           3          3     3           3           3               3   \n",
       "A*11:01           7          7     7           7           7               7   \n",
       "A*24:02          22         22    22          22          22              22   \n",
       "A*68:01           2          2     2           2           2               2   \n",
       "A*68:02           2          2     2           2           2               2   \n",
       "A*68:12           1          1     1           1           1               1   \n",
       "B*07:02           5          5     5           5           5               5   \n",
       "B*08:01          16         16    16          16          16              16   \n",
       "B*08:02           1          1     1           1           1               1   \n",
       "B*08:136          1          1     1           1           1               1   \n",
       "B*14:02           2          2     2           2           2               2   \n",
       "B*15:01           3          3     3           3           3               3   \n",
       "B*18:01           2          2     2           2           2               2   \n",
       "B*27:04           1          1     1           1           1               1   \n",
       "B*27:05          13         13    13          13          13              13   \n",
       "B*27:06           1          1     1           1           1               1   \n",
       "B*27:09           8          8     8           8           8               8   \n",
       "B*27:13           2          2     2           2           2               2   \n",
       "B*35:01           1          1     1           1           1               1   \n",
       "B*35:08          24         24    24          24          24              24   \n",
       "B*35:42          24         24    24          24          24              24   \n",
       "B*39:01           3          3     3           3           3               3   \n",
       "B*41:43           1          1     1           1           1               1   \n",
       "B*42:01           2          2     2           2           2               2   \n",
       "B*42:02           1          1     1           1           1               1   \n",
       "B*44:02           9          9     9           9           9               9   \n",
       "B*44:03           6          6     6           6           6               6   \n",
       "B*44:05           6          6     6           6           6               6   \n",
       "B*46:01           1          1     1           1           1               1   \n",
       "B*51:193          3          3     3           3           3               3   \n",
       "B*53:01           2          2     2           2           2               2   \n",
       "B*57:01          11         11    11          11          11              11   \n",
       "B*57:03           3          3     3           3           3               3   \n",
       "B*57:06           3          3     3           3           3               3   \n",
       "B*58:01           2          2     2           2           2               2   \n",
       "B*81:01           3          3     3           3           3               3   \n",
       "C*03:04           1          1     1           1           1               1   \n",
       "C*04:09N          2          2     2           2           2               2   \n",
       "C*08:22           1          1     1           1           1               1   \n",
       "\n",
       "                antigen_id  antigen_seq  antigen_len  antigen_name  \\\n",
       "a_chain_allele                                                       \n",
       "A*01:01                  6            6            6             6   \n",
       "A*02:01                186          186          186           101   \n",
       "A*02:03                  1            1            1             1   \n",
       "A*02:06                  1            1            1             1   \n",
       "A*02:07                  1            1            1             1   \n",
       "A*02:30                  1            1            1             1   \n",
       "A*02:562                 1            1            1             0   \n",
       "A*02:66                  1            1            1             0   \n",
       "A*03:01                  3            3            3             3   \n",
       "A*11:01                  7            7            7             4   \n",
       "A*24:02                 22           22           22            22   \n",
       "A*68:01                  2            2            2             2   \n",
       "A*68:02                  2            2            2             2   \n",
       "A*68:12                  1            1            1             1   \n",
       "B*07:02                  5            5            5             5   \n",
       "B*08:01                 16           16           16            15   \n",
       "B*08:02                  1            1            1             1   \n",
       "B*08:136                 1            1            1             1   \n",
       "B*14:02                  2            2            2             2   \n",
       "B*15:01                  3            3            3             3   \n",
       "B*18:01                  2            2            2             2   \n",
       "B*27:04                  1            1            1             1   \n",
       "B*27:05                 13           13           13            11   \n",
       "B*27:06                  1            1            1             1   \n",
       "B*27:09                  8            8            8             5   \n",
       "B*27:13                  2            2            2             2   \n",
       "B*35:01                  1            1            1             0   \n",
       "B*35:08                 24           24           24            17   \n",
       "B*35:42                 24           24           24            21   \n",
       "B*39:01                  3            3            3             3   \n",
       "B*41:43                  1            1            1             1   \n",
       "B*42:01                  2            2            2             2   \n",
       "B*42:02                  1            1            1             1   \n",
       "B*44:02                  9            9            9             3   \n",
       "B*44:03                  6            6            6             5   \n",
       "B*44:05                  6            6            6             4   \n",
       "B*46:01                  1            1            1             1   \n",
       "B*51:193                 3            3            3             3   \n",
       "B*53:01                  2            2            2             2   \n",
       "B*57:01                 11           11           11            10   \n",
       "B*57:03                  3            3            3             2   \n",
       "B*57:06                  3            3            3             3   \n",
       "B*58:01                  2            2            2             0   \n",
       "B*81:01                  3            3            3             3   \n",
       "C*03:04                  1            1            1             1   \n",
       "C*04:09N                 2            2            2             0   \n",
       "C*08:22                  1            1            1             1   \n",
       "\n",
       "                antigen_organism  comments  \n",
       "a_chain_allele                              \n",
       "A*01:01                        6         0  \n",
       "A*02:01                      101        19  \n",
       "A*02:03                        1         1  \n",
       "A*02:06                        1         1  \n",
       "A*02:07                        1         1  \n",
       "A*02:30                        1         0  \n",
       "A*02:562                       0         0  \n",
       "A*02:66                        0         1  \n",
       "A*03:01                        3         1  \n",
       "A*11:01                        4         2  \n",
       "A*24:02                       22         0  \n",
       "A*68:01                        2         1  \n",
       "A*68:02                        2         0  \n",
       "A*68:12                        1         0  \n",
       "B*07:02                        5         0  \n",
       "B*08:01                       15         0  \n",
       "B*08:02                        1         0  \n",
       "B*08:136                       1         0  \n",
       "B*14:02                        2         0  \n",
       "B*15:01                        3         0  \n",
       "B*18:01                        2         0  \n",
       "B*27:04                        1         0  \n",
       "B*27:05                       11         2  \n",
       "B*27:06                        1         0  \n",
       "B*27:09                        5         1  \n",
       "B*27:13                        2         0  \n",
       "B*35:01                        0         0  \n",
       "B*35:08                       17         1  \n",
       "B*35:42                       21         4  \n",
       "B*39:01                        3         0  \n",
       "B*41:43                        1         0  \n",
       "B*42:01                        2         0  \n",
       "B*42:02                        1         0  \n",
       "B*44:02                        3         0  \n",
       "B*44:03                        5         1  \n",
       "B*44:05                        4         0  \n",
       "B*46:01                        1         0  \n",
       "B*51:193                       3         0  \n",
       "B*53:01                        2         1  \n",
       "B*57:01                       10         0  \n",
       "B*57:03                        2         0  \n",
       "B*57:06                        3         0  \n",
       "B*58:01                        0         0  \n",
       "B*81:01                        3         0  \n",
       "C*03:04                        1         0  \n",
       "C*04:09N                       0         2  \n",
       "C*08:22                        1         0  "
      ]
     },
     "execution_count": 438,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iclass.groupby('a_chain_allele').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def peptides_rmsd(path1, chain1, path2, chain2):\n",
    "    cmd.reinitialize()\n",
    "    cmd.load(path1, 'obj1')\n",
    "    cmd.load(path2, 'obj2')\n",
    "    \n",
    "    res1 = []\n",
    "    space = {'ar' : res1}\n",
    "    cmd.iterate_state(0, 'obj1//%s//CA' % chain1, 'ar.append((resi, resn))', space=space)\n",
    "    cmd.alter('obj1//%s//' % chain1, 'segi=\"\"')\n",
    "    cmd.alter('obj1//%s//' % chain1, 'chain=\"XXX\"')\n",
    "    #print(res1)\n",
    "    \n",
    "    # change identifiers of obj2 CA to make them similar to obj1\n",
    "    resi2 = []\n",
    "    space = {'ar' : resi2}\n",
    "    cmd.iterate_state(0, 'obj2//%s//CA' % chain2, 'ar.append(resi)', space=space)\n",
    "    cmd.alter('obj2//%s//' % chain2, 'segi=\"\"')\n",
    "    cmd.alter('obj2//%s//' % chain2, 'chain=\"XXX\"')\n",
    "    for idx, newidx in zip(resi2, res1):\n",
    "        cmd.alter('obj2//XXX/%s/' % idx, 'resi=\"%s\"' % newidx[0])\n",
    "        cmd.alter('obj2//XXX/%s/' % idx, 'resn=\"%s\"' % newidx[1])\n",
    "        \n",
    "    rmsd = cmd.rms('obj1//XXX//CA', 'obj2//XXX//CA')\n",
    "\n",
    "    return rmsd\n",
    "    \n",
    "def rmsd_list(name, pdbtable, pep2pdb, mhc2pdb, maxnum=20):\n",
    "    # read info for molecule 'name'\n",
    "    line1 = pdbtable[pdbtable.pdb == name].iloc[0, :]\n",
    "    path1 = corrected_pdb_path+'/'+name+'.pdb'\n",
    "    chain1 = line1.antigen_id\n",
    "    allele1 = line1.a_chain_allele\n",
    "    pepseq1 = line1.antigen_seq\n",
    "    \n",
    "    # check that peptide is present in pep2pdb table\n",
    "    if pepseq1 not in pep2pdb.index:\n",
    "        print(name+': bad peptide')\n",
    "        return {name: [-1]*maxnum}\n",
    "    \n",
    "    # discard similar alleles\n",
    "    nonsimilar = pdbtable[pdbtable.a_chain_allele != allele1].pdb\n",
    "    \n",
    "    res = []\n",
    "    chunk = mhc2pdb.loc[mhc2pdb.index.str.startswith(allele1), :]\n",
    "    #best_pdbs = (chunk.apply(min) + pep2pdb.loc[pepseq1, :]/FACTOR)[nonsimilar].sort_values()\n",
    "    best_pdbs = (chunk.apply(min) + pep2pdb.loc[pepseq1, :]/FACTOR).sort_values()\n",
    "    for pdb in best_pdbs.index[:maxnum]:\n",
    "        if (best_pdbs[pdb] < np.inf):\n",
    "            line2 = pdbtable[pdbtable.pdb == pdb].iloc[0, :]\n",
    "            path2 = corrected_pdb_path+'/'+pdb+'.pdb'\n",
    "            chain2 = line2.antigen_id\n",
    "            #print(name+' vs '+pdb+': '+str(peptides_rmsd(path1, chain1, path2, chain2)))\n",
    "            res.append(peptides_rmsd(path1, chain1, path2, chain2))\n",
    "        else:\n",
    "            #print(name+' vs '+pdb+': do not match')\n",
    "            res.append(np.nan)\n",
    "    return {name: res}\n",
    "    \n",
    "#rmsd_dict = {}\n",
    "#for name in iclass.pdb:\n",
    "#    rmsd_dict.update(rmsd_list(name, iclass, peptides_vs_pdb, alleles_vs_pdb))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.77944862155388472"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum((pd.DataFrame(rmsd_dict).loc[:10, :].replace(-1.0, 10) < 1.0).apply(any))/float(len(rmsd_dict))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1a1m</th>\n",
       "      <th>1a1n</th>\n",
       "      <th>1a1o</th>\n",
       "      <th>1a9e</th>\n",
       "      <th>1agb</th>\n",
       "      <th>1agc</th>\n",
       "      <th>1agd</th>\n",
       "      <th>1age</th>\n",
       "      <th>1agf</th>\n",
       "      <th>1akj</th>\n",
       "      <th>...</th>\n",
       "      <th>5hgd</th>\n",
       "      <th>5hgh</th>\n",
       "      <th>5hhn</th>\n",
       "      <th>5hho</th>\n",
       "      <th>5hhp</th>\n",
       "      <th>5hhq</th>\n",
       "      <th>5inc</th>\n",
       "      <th>5ind</th>\n",
       "      <th>5iro</th>\n",
       "      <th>5swq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.139542</td>\n",
       "      <td>1.132532</td>\n",
       "      <td>1.327055</td>\n",
       "      <td>0.326978</td>\n",
       "      <td>2.086508e+00</td>\n",
       "      <td>2.166762e+00</td>\n",
       "      <td>2.155766e+00</td>\n",
       "      <td>2.121538e+00</td>\n",
       "      <td>2.107955e+00</td>\n",
       "      <td>1.337969</td>\n",
       "      <td>...</td>\n",
       "      <td>2.536710e+00</td>\n",
       "      <td>2.524906e+00</td>\n",
       "      <td>0.989251</td>\n",
       "      <td>1.538476</td>\n",
       "      <td>1.343414</td>\n",
       "      <td>1.096058</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.675522</td>\n",
       "      <td>1.148401</td>\n",
       "      <td>4.941041e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.206494</td>\n",
       "      <td>1.034485</td>\n",
       "      <td>1.320663</td>\n",
       "      <td>1.622864</td>\n",
       "      <td>5.807886e-01</td>\n",
       "      <td>4.132785e-01</td>\n",
       "      <td>4.071409e-01</td>\n",
       "      <td>3.949687e-01</td>\n",
       "      <td>3.758281e-01</td>\n",
       "      <td>1.135945</td>\n",
       "      <td>...</td>\n",
       "      <td>6.040261e-11</td>\n",
       "      <td>6.040261e-11</td>\n",
       "      <td>1.254031</td>\n",
       "      <td>0.983763</td>\n",
       "      <td>1.053181</td>\n",
       "      <td>1.222808</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.685045</td>\n",
       "      <td>1.143062</td>\n",
       "      <td>2.114923e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.092499</td>\n",
       "      <td>1.019194</td>\n",
       "      <td>0.460382</td>\n",
       "      <td>1.350060</td>\n",
       "      <td>6.051277e-01</td>\n",
       "      <td>6.595728e-01</td>\n",
       "      <td>5.924562e-01</td>\n",
       "      <td>5.908124e-01</td>\n",
       "      <td>6.035284e-01</td>\n",
       "      <td>1.207939</td>\n",
       "      <td>...</td>\n",
       "      <td>2.629005e+00</td>\n",
       "      <td>2.610812e+00</td>\n",
       "      <td>0.944017</td>\n",
       "      <td>0.863628</td>\n",
       "      <td>1.014962</td>\n",
       "      <td>1.062107</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.679242</td>\n",
       "      <td>0.583344</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.697405</td>\n",
       "      <td>2.097007</td>\n",
       "      <td>0.454211</td>\n",
       "      <td>1.342483</td>\n",
       "      <td>4.747235e-01</td>\n",
       "      <td>3.590615e-01</td>\n",
       "      <td>3.531187e-01</td>\n",
       "      <td>3.444836e-01</td>\n",
       "      <td>2.776660e-01</td>\n",
       "      <td>1.039592</td>\n",
       "      <td>...</td>\n",
       "      <td>2.575205e+00</td>\n",
       "      <td>2.561959e+00</td>\n",
       "      <td>0.920729</td>\n",
       "      <td>1.205438</td>\n",
       "      <td>1.103532</td>\n",
       "      <td>1.081025</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.651795</td>\n",
       "      <td>1.109387</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.703075</td>\n",
       "      <td>1.028739</td>\n",
       "      <td>1.350060</td>\n",
       "      <td>0.957108</td>\n",
       "      <td>1.062962e+00</td>\n",
       "      <td>9.023880e-01</td>\n",
       "      <td>9.070487e-01</td>\n",
       "      <td>8.877627e-01</td>\n",
       "      <td>1.155452e+00</td>\n",
       "      <td>0.726095</td>\n",
       "      <td>...</td>\n",
       "      <td>2.659719e+00</td>\n",
       "      <td>2.648761e+00</td>\n",
       "      <td>0.368661</td>\n",
       "      <td>0.522089</td>\n",
       "      <td>0.472819</td>\n",
       "      <td>0.492953</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.409153</td>\n",
       "      <td>1.224504</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1.622864</td>\n",
       "      <td>1.241411</td>\n",
       "      <td>0.476393</td>\n",
       "      <td>1.255134</td>\n",
       "      <td>8.332888e-01</td>\n",
       "      <td>7.382293e-01</td>\n",
       "      <td>7.284047e-01</td>\n",
       "      <td>7.574609e-01</td>\n",
       "      <td>7.270477e-01</td>\n",
       "      <td>0.807407</td>\n",
       "      <td>...</td>\n",
       "      <td>2.566694e+00</td>\n",
       "      <td>2.555122e+00</td>\n",
       "      <td>0.560062</td>\n",
       "      <td>0.866106</td>\n",
       "      <td>0.684999</td>\n",
       "      <td>0.668898</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.680090</td>\n",
       "      <td>0.581473</td>\n",
       "      <td>2.898664e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1.218730</td>\n",
       "      <td>0.887763</td>\n",
       "      <td>0.463680</td>\n",
       "      <td>1.260010</td>\n",
       "      <td>5.329272e-01</td>\n",
       "      <td>4.068665e-01</td>\n",
       "      <td>4.070600e-01</td>\n",
       "      <td>4.201115e-01</td>\n",
       "      <td>3.662108e-01</td>\n",
       "      <td>0.618642</td>\n",
       "      <td>...</td>\n",
       "      <td>2.271165e+00</td>\n",
       "      <td>2.283606e+00</td>\n",
       "      <td>0.271339</td>\n",
       "      <td>0.673869</td>\n",
       "      <td>0.409925</td>\n",
       "      <td>0.269813</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.588419</td>\n",
       "      <td>1.208005</td>\n",
       "      <td>1.717771e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1.386504</td>\n",
       "      <td>0.902388</td>\n",
       "      <td>0.468978</td>\n",
       "      <td>1.429574</td>\n",
       "      <td>6.367482e-01</td>\n",
       "      <td>6.869991e-01</td>\n",
       "      <td>6.853245e-01</td>\n",
       "      <td>6.887801e-01</td>\n",
       "      <td>8.408865e-01</td>\n",
       "      <td>0.951957</td>\n",
       "      <td>...</td>\n",
       "      <td>2.600358e+00</td>\n",
       "      <td>2.586132e+00</td>\n",
       "      <td>0.539762</td>\n",
       "      <td>0.917931</td>\n",
       "      <td>0.742238</td>\n",
       "      <td>0.640306</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.447534</td>\n",
       "      <td>0.574689</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1.171789</td>\n",
       "      <td>1.062962</td>\n",
       "      <td>0.454285</td>\n",
       "      <td>0.890916</td>\n",
       "      <td>6.040261e-11</td>\n",
       "      <td>6.040261e-11</td>\n",
       "      <td>6.040261e-11</td>\n",
       "      <td>6.040261e-11</td>\n",
       "      <td>6.040261e-11</td>\n",
       "      <td>1.680874</td>\n",
       "      <td>...</td>\n",
       "      <td>2.603833e+00</td>\n",
       "      <td>2.590298e+00</td>\n",
       "      <td>1.594993</td>\n",
       "      <td>1.893710</td>\n",
       "      <td>1.617371</td>\n",
       "      <td>1.567389</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.604360</td>\n",
       "      <td>1.212600</td>\n",
       "      <td>6.040261e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1.180262</td>\n",
       "      <td>0.907049</td>\n",
       "      <td>0.800505</td>\n",
       "      <td>0.468812</td>\n",
       "      <td>5.766785e-01</td>\n",
       "      <td>4.084007e-01</td>\n",
       "      <td>3.551152e-01</td>\n",
       "      <td>3.821214e-01</td>\n",
       "      <td>5.653157e-01</td>\n",
       "      <td>0.971857</td>\n",
       "      <td>...</td>\n",
       "      <td>2.070001e+00</td>\n",
       "      <td>2.053192e+00</td>\n",
       "      <td>0.721134</td>\n",
       "      <td>1.150041</td>\n",
       "      <td>0.867111</td>\n",
       "      <td>0.665036</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.647371</td>\n",
       "      <td>0.541458</td>\n",
       "      <td>3.308478e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1.757656</td>\n",
       "      <td>1.163836</td>\n",
       "      <td>1.587393</td>\n",
       "      <td>0.535801</td>\n",
       "      <td>8.487645e-01</td>\n",
       "      <td>8.567593e-01</td>\n",
       "      <td>8.131823e-01</td>\n",
       "      <td>8.689265e-01</td>\n",
       "      <td>9.007719e-01</td>\n",
       "      <td>1.066843</td>\n",
       "      <td>...</td>\n",
       "      <td>1.927479e+00</td>\n",
       "      <td>1.911711e+00</td>\n",
       "      <td>0.910234</td>\n",
       "      <td>0.949056</td>\n",
       "      <td>0.977458</td>\n",
       "      <td>1.064711</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.474176</td>\n",
       "      <td>1.160619</td>\n",
       "      <td>6.040261e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1.929433</td>\n",
       "      <td>1.155452</td>\n",
       "      <td>1.633649</td>\n",
       "      <td>1.501266</td>\n",
       "      <td>6.089846e-01</td>\n",
       "      <td>4.083173e-01</td>\n",
       "      <td>3.859410e-01</td>\n",
       "      <td>4.254408e-01</td>\n",
       "      <td>6.788516e-01</td>\n",
       "      <td>1.790418</td>\n",
       "      <td>...</td>\n",
       "      <td>2.014124e+00</td>\n",
       "      <td>2.012628e+00</td>\n",
       "      <td>1.714468</td>\n",
       "      <td>1.916640</td>\n",
       "      <td>1.799217</td>\n",
       "      <td>1.823683</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.635783</td>\n",
       "      <td>1.146211</td>\n",
       "      <td>6.040261e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1.647371</td>\n",
       "      <td>0.812329</td>\n",
       "      <td>1.474176</td>\n",
       "      <td>0.490927</td>\n",
       "      <td>5.406336e-01</td>\n",
       "      <td>3.907453e-01</td>\n",
       "      <td>3.583325e-01</td>\n",
       "      <td>4.041180e-01</td>\n",
       "      <td>6.338676e-01</td>\n",
       "      <td>0.923033</td>\n",
       "      <td>...</td>\n",
       "      <td>1.898696e+00</td>\n",
       "      <td>1.875761e+00</td>\n",
       "      <td>0.495135</td>\n",
       "      <td>0.795908</td>\n",
       "      <td>0.639512</td>\n",
       "      <td>0.581785</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.518063</td>\n",
       "      <td>1.104014</td>\n",
       "      <td>6.040261e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1.519318</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.134484</td>\n",
       "      <td>0.598557</td>\n",
       "      <td>5.615574e-01</td>\n",
       "      <td>4.053616e-01</td>\n",
       "      <td>3.550689e-01</td>\n",
       "      <td>3.808718e-01</td>\n",
       "      <td>5.876871e-01</td>\n",
       "      <td>0.996804</td>\n",
       "      <td>...</td>\n",
       "      <td>2.656981e+00</td>\n",
       "      <td>2.662973e+00</td>\n",
       "      <td>0.618438</td>\n",
       "      <td>0.928123</td>\n",
       "      <td>0.831143</td>\n",
       "      <td>0.784584</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.630134</td>\n",
       "      <td>1.179902</td>\n",
       "      <td>6.040261e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1.369379</td>\n",
       "      <td>0.956770</td>\n",
       "      <td>0.855238</td>\n",
       "      <td>0.635055</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>1.559717</td>\n",
       "      <td>...</td>\n",
       "      <td>2.597275e+00</td>\n",
       "      <td>2.583622e+00</td>\n",
       "      <td>1.608897</td>\n",
       "      <td>1.660735</td>\n",
       "      <td>1.578195</td>\n",
       "      <td>1.613650</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.617267</td>\n",
       "      <td>0.750980</td>\n",
       "      <td>6.040261e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1.470852</td>\n",
       "      <td>1.482900</td>\n",
       "      <td>0.875594</td>\n",
       "      <td>0.778019</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>0.977322</td>\n",
       "      <td>...</td>\n",
       "      <td>2.512449e+00</td>\n",
       "      <td>2.494879e+00</td>\n",
       "      <td>0.554949</td>\n",
       "      <td>0.478303</td>\n",
       "      <td>0.657773</td>\n",
       "      <td>0.666630</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.509828</td>\n",
       "      <td>1.049577</td>\n",
       "      <td>1.886049e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>0.589384</td>\n",
       "      <td>0.960989</td>\n",
       "      <td>0.863809</td>\n",
       "      <td>0.636216</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>0.760573</td>\n",
       "      <td>...</td>\n",
       "      <td>2.181383e+00</td>\n",
       "      <td>2.192470e+00</td>\n",
       "      <td>0.580481</td>\n",
       "      <td>0.904888</td>\n",
       "      <td>0.745165</td>\n",
       "      <td>0.707606</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.527282</td>\n",
       "      <td>1.227514</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>0.615671</td>\n",
       "      <td>0.902071</td>\n",
       "      <td>0.858423</td>\n",
       "      <td>0.886959</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>0.952686</td>\n",
       "      <td>...</td>\n",
       "      <td>2.620124e+00</td>\n",
       "      <td>2.606464e+00</td>\n",
       "      <td>0.655000</td>\n",
       "      <td>0.844059</td>\n",
       "      <td>0.852035</td>\n",
       "      <td>0.823563</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.573882</td>\n",
       "      <td>1.098909</td>\n",
       "      <td>6.040261e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>0.607808</td>\n",
       "      <td>0.936566</td>\n",
       "      <td>0.768565</td>\n",
       "      <td>1.487055</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>1.093197</td>\n",
       "      <td>...</td>\n",
       "      <td>2.664209e+00</td>\n",
       "      <td>2.649667e+00</td>\n",
       "      <td>0.606182</td>\n",
       "      <td>0.632938</td>\n",
       "      <td>0.736660</td>\n",
       "      <td>0.746676</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.549612</td>\n",
       "      <td>0.792180</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>0.620960</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>0.448787</td>\n",
       "      <td>1.567465</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>0.966652</td>\n",
       "      <td>...</td>\n",
       "      <td>2.588381e+00</td>\n",
       "      <td>2.576182e+00</td>\n",
       "      <td>0.512605</td>\n",
       "      <td>0.931682</td>\n",
       "      <td>0.719955</td>\n",
       "      <td>0.631892</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.602342</td>\n",
       "      <td>1.207892</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20 rows  399 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        1a1m      1a1n      1a1o      1a9e          1agb          1agc  \\\n",
       "0   1.139542  1.132532  1.327055  0.326978  2.086508e+00  2.166762e+00   \n",
       "1   1.206494  1.034485  1.320663  1.622864  5.807886e-01  4.132785e-01   \n",
       "2   1.092499  1.019194  0.460382  1.350060  6.051277e-01  6.595728e-01   \n",
       "3   1.697405  2.097007  0.454211  1.342483  4.747235e-01  3.590615e-01   \n",
       "4   1.703075  1.028739  1.350060  0.957108  1.062962e+00  9.023880e-01   \n",
       "5   1.622864  1.241411  0.476393  1.255134  8.332888e-01  7.382293e-01   \n",
       "6   1.218730  0.887763  0.463680  1.260010  5.329272e-01  4.068665e-01   \n",
       "7   1.386504  0.902388  0.468978  1.429574  6.367482e-01  6.869991e-01   \n",
       "8   1.171789  1.062962  0.454285  0.890916  6.040261e-11  6.040261e-11   \n",
       "9   1.180262  0.907049  0.800505  0.468812  5.766785e-01  4.084007e-01   \n",
       "10  1.757656  1.163836  1.587393  0.535801  8.487645e-01  8.567593e-01   \n",
       "11  1.929433  1.155452  1.633649  1.501266  6.089846e-01  4.083173e-01   \n",
       "12  1.647371  0.812329  1.474176  0.490927  5.406336e-01  3.907453e-01   \n",
       "13  1.519318  0.000000  1.134484  0.598557  5.615574e-01  4.053616e-01   \n",
       "14  1.369379  0.956770  0.855238  0.635055 -1.000000e+00 -1.000000e+00   \n",
       "15  1.470852  1.482900  0.875594  0.778019 -1.000000e+00 -1.000000e+00   \n",
       "16  0.589384  0.960989  0.863809  0.636216 -1.000000e+00 -1.000000e+00   \n",
       "17  0.615671  0.902071  0.858423  0.886959 -1.000000e+00 -1.000000e+00   \n",
       "18  0.607808  0.936566  0.768565  1.487055 -1.000000e+00 -1.000000e+00   \n",
       "19  0.620960 -1.000000  0.448787  1.567465 -1.000000e+00 -1.000000e+00   \n",
       "\n",
       "            1agd          1age          1agf      1akj      ...       \\\n",
       "0   2.155766e+00  2.121538e+00  2.107955e+00  1.337969      ...        \n",
       "1   4.071409e-01  3.949687e-01  3.758281e-01  1.135945      ...        \n",
       "2   5.924562e-01  5.908124e-01  6.035284e-01  1.207939      ...        \n",
       "3   3.531187e-01  3.444836e-01  2.776660e-01  1.039592      ...        \n",
       "4   9.070487e-01  8.877627e-01  1.155452e+00  0.726095      ...        \n",
       "5   7.284047e-01  7.574609e-01  7.270477e-01  0.807407      ...        \n",
       "6   4.070600e-01  4.201115e-01  3.662108e-01  0.618642      ...        \n",
       "7   6.853245e-01  6.887801e-01  8.408865e-01  0.951957      ...        \n",
       "8   6.040261e-11  6.040261e-11  6.040261e-11  1.680874      ...        \n",
       "9   3.551152e-01  3.821214e-01  5.653157e-01  0.971857      ...        \n",
       "10  8.131823e-01  8.689265e-01  9.007719e-01  1.066843      ...        \n",
       "11  3.859410e-01  4.254408e-01  6.788516e-01  1.790418      ...        \n",
       "12  3.583325e-01  4.041180e-01  6.338676e-01  0.923033      ...        \n",
       "13  3.550689e-01  3.808718e-01  5.876871e-01  0.996804      ...        \n",
       "14 -1.000000e+00 -1.000000e+00 -1.000000e+00  1.559717      ...        \n",
       "15 -1.000000e+00 -1.000000e+00 -1.000000e+00  0.977322      ...        \n",
       "16 -1.000000e+00 -1.000000e+00 -1.000000e+00  0.760573      ...        \n",
       "17 -1.000000e+00 -1.000000e+00 -1.000000e+00  0.952686      ...        \n",
       "18 -1.000000e+00 -1.000000e+00 -1.000000e+00  1.093197      ...        \n",
       "19 -1.000000e+00 -1.000000e+00 -1.000000e+00  0.966652      ...        \n",
       "\n",
       "            5hgd          5hgh      5hhn      5hho      5hhp      5hhq  5inc  \\\n",
       "0   2.536710e+00  2.524906e+00  0.989251  1.538476  1.343414  1.096058    -1   \n",
       "1   6.040261e-11  6.040261e-11  1.254031  0.983763  1.053181  1.222808    -1   \n",
       "2   2.629005e+00  2.610812e+00  0.944017  0.863628  1.014962  1.062107    -1   \n",
       "3   2.575205e+00  2.561959e+00  0.920729  1.205438  1.103532  1.081025    -1   \n",
       "4   2.659719e+00  2.648761e+00  0.368661  0.522089  0.472819  0.492953    -1   \n",
       "5   2.566694e+00  2.555122e+00  0.560062  0.866106  0.684999  0.668898    -1   \n",
       "6   2.271165e+00  2.283606e+00  0.271339  0.673869  0.409925  0.269813    -1   \n",
       "7   2.600358e+00  2.586132e+00  0.539762  0.917931  0.742238  0.640306    -1   \n",
       "8   2.603833e+00  2.590298e+00  1.594993  1.893710  1.617371  1.567389    -1   \n",
       "9   2.070001e+00  2.053192e+00  0.721134  1.150041  0.867111  0.665036    -1   \n",
       "10  1.927479e+00  1.911711e+00  0.910234  0.949056  0.977458  1.064711    -1   \n",
       "11  2.014124e+00  2.012628e+00  1.714468  1.916640  1.799217  1.823683    -1   \n",
       "12  1.898696e+00  1.875761e+00  0.495135  0.795908  0.639512  0.581785    -1   \n",
       "13  2.656981e+00  2.662973e+00  0.618438  0.928123  0.831143  0.784584    -1   \n",
       "14  2.597275e+00  2.583622e+00  1.608897  1.660735  1.578195  1.613650    -1   \n",
       "15  2.512449e+00  2.494879e+00  0.554949  0.478303  0.657773  0.666630    -1   \n",
       "16  2.181383e+00  2.192470e+00  0.580481  0.904888  0.745165  0.707606    -1   \n",
       "17  2.620124e+00  2.606464e+00  0.655000  0.844059  0.852035  0.823563    -1   \n",
       "18  2.664209e+00  2.649667e+00  0.606182  0.632938  0.736660  0.746676    -1   \n",
       "19  2.588381e+00  2.576182e+00  0.512605  0.931682  0.719955  0.631892    -1   \n",
       "\n",
       "        5ind      5iro          5swq  \n",
       "0   1.675522  1.148401  4.941041e-02  \n",
       "1   1.685045  1.143062  2.114923e-01  \n",
       "2   1.679242  0.583344  0.000000e+00  \n",
       "3   1.651795  1.109387  0.000000e+00  \n",
       "4   1.409153  1.224504  0.000000e+00  \n",
       "5   1.680090  0.581473  2.898664e-01  \n",
       "6   1.588419  1.208005  1.717771e-01  \n",
       "7   1.447534  0.574689  0.000000e+00  \n",
       "8   1.604360  1.212600  6.040261e-11  \n",
       "9   1.647371  0.541458  3.308478e-01  \n",
       "10  1.474176  1.160619  6.040261e-11  \n",
       "11  0.635783  1.146211  6.040261e-11  \n",
       "12  1.518063  1.104014  6.040261e-11  \n",
       "13  0.630134  1.179902  6.040261e-11  \n",
       "14  1.617267  0.750980  6.040261e-11  \n",
       "15  1.509828  1.049577  1.886049e-01  \n",
       "16  1.527282  1.227514  0.000000e+00  \n",
       "17  1.573882  1.098909  6.040261e-11  \n",
       "18  1.549612  0.792180  0.000000e+00  \n",
       "19  1.602342  1.207892  0.000000e+00  \n",
       "\n",
       "[20 rows x 399 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rmsd_table = pd.read_csv('../sandbox/peptide_rmsd_nonred', sep='\\t')\n",
    "rmsd_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "t1 = pd.read_csv('/home/mikhail/Projects/SAXS/libsaxs/examples/benchmark/Real_new/piper/2CFU/prms/rot70k.0.0.4.prm', \n",
    "                 sep='\\s+', header=None)\n",
    "t2 = pd.read_csv('/home/mikhail/Projects/SAXS/libsaxs/examples/benchmark/Real_new/piper/2CFU/prms/rot70k.mol2', \n",
    "                 sep='\\s+', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "t1 = pd.read_csv('/home/mikhail/Projects/DOCK/libsaxs/complex_testing/benchmark/Real_new/5T1O/saxs_15_20.000000_51.000000_0.2', \n",
    "                 sep='\\s+', header=None, skiprows=1)\n",
    "t1.sort_values(1).iloc[:10000, :].to_csv('/home/mikhail/Projects/DOCK/libsaxs/complex_testing/benchmark/Real_new/5T1O/saxs_10000', \n",
    "                 sep='\\t', header=None, index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda env:python2]",
   "language": "python",
   "name": "conda-env-python2-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
