{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import Bio.PDB\n",
    "import Bio.SeqIO\n",
    "from Bio import pairwise2\n",
    "from Bio.SubsMat import MatrixInfo as matlist\n",
    "from Bio.SeqUtils import seq3\n",
    "\n",
    "import re\n",
    "import glob\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import subprocess\n",
    "\n",
    "import sys\n",
    "reload(sys)\n",
    "sys.setdefaultencoding('utf-8')\n",
    "\n",
    "# paths\n",
    "\n",
    "dbfiles = glob.glob(\"../data/mhc_alleles/hla/*prot.fasta\") + glob.glob(\"../data/mhc_alleles/mhc/*/*prot.fasta\")\n",
    "hla_fasta = '../data/mhc_alleles/hla/hla_prot.fasta'\n",
    "mhc_db_path = '../sandbox/db.fasta'\n",
    "epitope_db_path = '../data/epitopes/epitope_table.csv'\n",
    "blast_output_path = '../sandbox/blast_output'\n",
    "epitope_annot = '../sandbox/epitope_annot'\n",
    "mhc_annot = '../sandbox/mhc_annot'\n",
    "query_path = '../sandbox/query.fasta'\n",
    "pdb_path = '../data/structures'\n",
    "epitope_query_path = '../sandbox/epitope_query.fasta'\n",
    "badlist = '../data/structures/badlist'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import __main__\n",
    "__main__.pymol_argv = ['pymol','-Qc'] # Pymol: quiet and no GUI\n",
    "from time import sleep\n",
    "import pymol\n",
    "from pymol import cmd\n",
    "\n",
    "pymol.finish_launching()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# replace ' ' in indentifiers with ';'\n",
    "db = open(mhc_db_path, 'w')\n",
    "for dbfile in dbfiles:\n",
    "    f = open(dbfile, 'r')\n",
    "    text = f.read()\n",
    "    f.close()\n",
    "    db.write(';'.join(text.split(' ')))\n",
    "\n",
    "db.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# fetch all chains from pdbs, combine in one query file and blast against hla db\n",
    "\n",
    "def blast_pdb():\n",
    "    parser = Bio.PDB.PDBParser()\n",
    "    ppb = Bio.PDB.CaPPBuilder()\n",
    "\n",
    "    query_file = open(query_path, 'w')\n",
    "\n",
    "    pdbfiles = glob.glob(pdb_path + \"/*.pdb\")\n",
    "    for pdb in pdbfiles:\n",
    "        if pdb in pd.read_csv(badlist, sep='\\t', header=None)[0]:\n",
    "            continue\n",
    "            \n",
    "        structure = parser.get_structure(id=pdb[-8:-4], file=pdb)\n",
    "        for chain in structure[0]:\n",
    "            try:\n",
    "                query_file.write('>%s;%s\\n%s\\n' % (structure.id, chain.id, ''.join([str(s.get_sequence()) for s in ppb.build_peptides(chain)])))\n",
    "            except Exception:\n",
    "                continue\n",
    "\n",
    "    query_file.close()\n",
    "    \n",
    "    # submit query file for blast search\n",
    "    command = [\"blastp\", \"-query\", query_path, \n",
    "                     \"-subject\", mhc_db_path, \"-out\", \n",
    "                     blast_output_path, \"-outfmt\", str(6), \"-num_alignments\", str(1)]\n",
    "    if subprocess.call(command) != 0:\n",
    "        print('Failed to blast')\n",
    "        exit()\n",
    "    \n",
    "#blast_pdb()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# convert blast output to mhc table\n",
    "\n",
    "def host_n_allele(s):\n",
    "    host = s\n",
    "    if ';' in s:\n",
    "            host = s.split(';')[1]\n",
    "    if '-' in host:\n",
    "        return host.split('-')\n",
    "    else:\n",
    "        return ['Human', host]\n",
    "\n",
    "def make_mhc_table():\n",
    "    blasted = pd.read_csv(blast_output_path, sep='\\t', header=None)\n",
    "    pdb_list = list(set([x.split(';')[0] for x in blasted[0]]))\n",
    "    \n",
    "    mhc_table = []\n",
    "\n",
    "    for pdb in pdb_list:\n",
    "        piece = blasted[blasted[0].str.startswith(pdb)]\n",
    "        if piece.shape[0] < 2:\n",
    "            continue\n",
    "        thres = np.mean(piece[11])\n",
    "        chunk = piece[piece[11] > thres]\n",
    "        for i in chunk.index:\n",
    "            line = chunk.loc[i, :]\n",
    "            if line[2] < 90:\n",
    "                continue\n",
    "            hna = host_n_allele(line[1])\n",
    "            susp = 0\n",
    "            if line[11] < thres or line[2] < 95:\n",
    "                susp = 1\n",
    "            row = [line[0].split(';')[0], line[0].split(';')[1], 'class', hna[0], hna[1], susp]\n",
    "            mhc_table.append(row)\n",
    "\n",
    "    mhc_table = pd.DataFrame(mhc_table, columns=['pdb', 'chain', 'mhc_class', 'mhc_host', 'mhc_allele', 'unsure'])\n",
    "    mhc_table.to_csv(mhc_annot, sep='\\t', index=False)\n",
    "    \n",
    "#make_mhc_table()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# make epitope table using epitope database\n",
    "\n",
    "def make_epitope_table():\n",
    "    epitope_db = pd.read_csv(epitope_db_path, header=None, skiprows=1, low_memory=False)\n",
    "    epitope_table = []\n",
    "\n",
    "    query_file = open(query_path, 'r')\n",
    "    text = query_file.read()\n",
    "    query_file.close()\n",
    "    text = text.split()\n",
    "\n",
    "    epitope_query_file = open(epitope_query_path, 'w')\n",
    "    for i in range(len(text)):\n",
    "        if text[i][0] == '>':\n",
    "            if len(text[i+1]) < 30:\n",
    "                seq = text[i+1]\n",
    "                name = text[i].split(';')\n",
    "                epitope_query_file.write('%s\\n%s\\n' % (text[i], seq))\n",
    "                epitope_db_tmp = epitope_db[epitope_db[2].apply(lambda x: len(x)) == len(seq)]\n",
    "                epitope_db_tmp = epitope_db_tmp[epitope_db_tmp[2].str.contains(seq)]\n",
    "                trial = epitope_db_tmp[epitope_db_tmp[2] == seq]\n",
    "                if trial.shape[0] > 0:\n",
    "                    epitope_db_tmp = trial\n",
    "                if epitope_db_tmp.shape[0] > 0:\n",
    "                    row = [name[0][1:], name[1], seq, len(seq), epitope_db_tmp.iloc[0, 7], \n",
    "                           epitope_db_tmp.iloc[0, 9], epitope_db_tmp.iloc[0, 11]]\n",
    "                else:\n",
    "                    row = [name[0][1:], name[1], seq, len(seq), 'NaN', 'NaN', 'NaN']\n",
    "                epitope_table.append(row)\n",
    "\n",
    "    epitope_query_file.close()\n",
    "\n",
    "    epitope_table = pd.DataFrame(epitope_table, columns=['pdb', 'chain', 'sequence', 'length', 'name', 'organism', 'comments'])\n",
    "    epitope_table.to_csv(epitope_annot, sep='\\t', index=False)\n",
    "\n",
    "#make_epitope_table()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "epitope_table = pd.read_csv(epitope_annot, sep='\\t')\n",
    "mhc_table = pd.read_csv(mhc_annot, sep='\\t')\n",
    "\n",
    "mhc_table = mhc_table[mhc_table.mhc_host == 'Human']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# make detailed mhc table\n",
    "\n",
    "# check if chains are less than 8 A separated\n",
    "def chain_close(a, b):\n",
    "    mindist = 100\n",
    "    for a1 in a.get_atoms():\n",
    "        if mindist < 8:\n",
    "            return True\n",
    "        for a2 in b.get_atoms():\n",
    "            dist = a1-a2\n",
    "            if dist < mindist:\n",
    "                mindist = dist\n",
    "    return False\n",
    "\n",
    "# mean distance\n",
    "def chain_dist(a, b):\n",
    "    ave = 0\n",
    "    num = 0\n",
    "    for a1 in a.get_atoms():\n",
    "        mindist = 100\n",
    "        for a2 in b.get_atoms():\n",
    "            dist = a1-a2\n",
    "            if dist < mindist:\n",
    "                mindist = dist\n",
    "        ave += mindist\n",
    "        num += 1\n",
    "    return ave/num\n",
    "\n",
    "# pair MHC chains based on allele names and interchain distances\n",
    "def pair_chains(mhc, epi, alleles, pdb):\n",
    "    model = parser.get_structure(id=pdb[-8:-4], file=pdb)[0]\n",
    "    out = []\n",
    "    \n",
    "    for epi_ch in epi:\n",
    "        firmindist = 100\n",
    "        first_id = first_al = ''\n",
    "        for mhc_ch, allele in zip(mhc, alleles):\n",
    "            dist = chain_dist(model[epi_ch], model[mhc_ch])\n",
    "            if dist < firmindist:\n",
    "                firmindist = dist\n",
    "                first_id = mhc_ch\n",
    "                first_al = allele\n",
    "        \n",
    "        if len(allele.split('*')[0]) > 2:\n",
    "            second_id = second_al = ''\n",
    "            secmindist = 100\n",
    "            for mhc_ch, allele in zip(mhc, alleles):\n",
    "                if first_id == mhc_ch: \n",
    "                    continue\n",
    "                dist = chain_dist(model[epi_ch], model[mhc_ch])\n",
    "                if dist < secmindist:\n",
    "                    secmindist = dist\n",
    "                    second_id = mhc_ch\n",
    "                    second_al = allele\n",
    "            if (firmindist < 8 and secmindist < 8 and chain_close(model[first_id], model[second_id])):\n",
    "                if (first_al.split('*')[0][2] == 'B' and second_al.split('*')[0][2] == 'A'):\n",
    "                    out.append([[second_id, first_id], epi_ch])\n",
    "                else:\n",
    "                    out.append([[first_id, second_id], epi_ch])\n",
    "        else:\n",
    "            out.append([[first_id], epi_ch])\n",
    "    return out\n",
    "\n",
    "# verify no intersection\n",
    "def correctness(pairs):\n",
    "    for i in range(len(pairs)-1):\n",
    "        for j in range(i+1, len(pairs)):\n",
    "            for l in range(len(pairs[i][0])):\n",
    "                for m in range(len(pairs[j][0])):\n",
    "                    assert(pairs[i][0][l] != pairs[j][0][m])\n",
    "\n",
    "\n",
    "def make_annotation_table():\n",
    "    parser = Bio.PDB.PDBParser(QUIET=True)\n",
    "    ppb = Bio.PDB.CaPPBuilder()                    \n",
    "    pdb_set = list(set(mhc_table.pdb))\n",
    "    annotation = []\n",
    "\n",
    "    for pdb in pdb_set:\n",
    "        mhc_chunk = mhc_table[mhc_table.pdb == pdb]\n",
    "        epi_chunk = epitope_table[epitope_table.pdb == pdb]\n",
    "        pairs = pair_chains(list(mhc_chunk.chain), list(epi_chunk.chain), list(mhc_chunk.mhc_allele), \n",
    "                            \"%s/%s.pdb\" % (pdb_path, pdb))\n",
    "        correctness(pairs)\n",
    "\n",
    "        for pair in pairs:\n",
    "            line = ''\n",
    "            if len(pair[0]) == 1:\n",
    "                mhc_local = mhc_chunk[mhc_chunk.chain == pair[0][0]]\n",
    "                assert(mhc_local.iloc[0, 4][1] == '*')\n",
    "                #               pdb          class         host              1 chain  \n",
    "                line = [mhc_local.iloc[0, 0], 'I', mhc_local.iloc[0, 3],  mhc_local.iloc[0, 1], \n",
    "                #         1 chain allele       2 chain   2 chain allele      \n",
    "                        mhc_local.iloc[0, 4],   '',         'B2M']\n",
    "\n",
    "            if len(pair[0]) == 2:\n",
    "                mhc_a = mhc_chunk[mhc_chunk.chain == pair[0][0]]\n",
    "                mhc_b = mhc_chunk[mhc_chunk.chain == pair[0][1]]\n",
    "\n",
    "                assert(mhc_a.iloc[0, 4][:2] == mhc_b.iloc[0, 4][:2])\n",
    "                assert(mhc_a.iloc[0, 4][2] == 'A')\n",
    "                assert(mhc_b.iloc[0, 4][2] == 'B')\n",
    "                line = [mhc_chunk.iloc[0, 0], 'II', mhc_chunk.iloc[0, 3], mhc_a.iloc[0, 1], \n",
    "                        mhc_a.iloc[0, 4], mhc_b.iloc[0, 1], mhc_b.iloc[0, 4]]\n",
    "\n",
    "            line = line + list(epi_chunk[epi_chunk.chain == pair[1]].iloc[0, 1:])\n",
    "        annotation.append(line)\n",
    "\n",
    "        cols = ['pdb', 'mhc_class', 'host', 'a_chain_id', 'a_chain_allele', 'b_chain_id', 'b_chain_allele', \n",
    "            'antigen_id', 'antigen_seq', 'antigen_len', 'antigen_name', 'antigen_organism', 'comments']\n",
    "        annotation = pd.DataFrame(annotation, columns=cols).drop_duplicates()\n",
    "        annotation.to_csv('../sandbox/detailed_mhc', index=None, sep='\\t')\n",
    "        \n",
    "#make_annotation_table()\n",
    "\n",
    "annotation = pd.read_csv('../sandbox/detailed_mhc', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# define microglobuline chains\n",
    "def b2m_align(s):\n",
    "    b2m = 'MSRSVALAVLALLSLSGLEAIQRTPKIQVYSRHPAENGKSNFLNCYVSGFHPSDIEVDLLKNGERIEKDEYACRVNHVTLSQPKIVKWDRDM'\n",
    "    matrix = matlist.blosum90\n",
    "    gap_open = -1.0\n",
    "    gap_extend = -0.5\n",
    "    alns = pairwise2.align.globalds(s, b2m, matrix, gap_open, gap_extend)\n",
    "    return alns[0]\n",
    "\n",
    "def b2m_define(table):\n",
    "    chunk = table[table.mhc_class == 'I']\n",
    "    parser = Bio.PDB.PDBParser(QUIET=True)\n",
    "    ppb = Bio.PDB.CaPPBuilder() \n",
    "    \n",
    "    for i in chunk.index:\n",
    "        pdb = chunk.loc[i, 'pdb']\n",
    "        print(pdb)\n",
    "        structure = parser.get_structure(id=pdb, file=pdb_path+'/'+pdb+'.pdb')[0]\n",
    "        mhc = chunk.loc[i, 'a_chain_id']\n",
    "        pep = chunk.loc[i, 'antigen_id']\n",
    "        mindist = 1000\n",
    "        b2m = None;\n",
    "        for chain in structure.get_chains():\n",
    "            if chain.get_id() != mhc and chain.get_id() != pep:\n",
    "                if len(list(structure.get_chains())) == 3:\n",
    "                    b2m = chain\n",
    "                else:\n",
    "                    chain_seq = ''.join([str(x.get_sequence()) for x in ppb.build_peptides(chain)])\n",
    "                    if b2m_align(chain_seq)[2] > 400:\n",
    "                        if chain_close(chain, structure[mhc]):\n",
    "                            dist = chain_dist(chain, structure[mhc])\n",
    "                            if dist <= mindist:\n",
    "                                mindist = dist\n",
    "                                b2m = chain\n",
    "                                \n",
    "        if b2m == None:\n",
    "            table.loc[i, 'b_chain_id'] = None\n",
    "            print('bsm chain = %s' % 'None')\n",
    "        else:\n",
    "            table.loc[i, 'b_chain_id'] = b2m.get_id()\n",
    "            print('bsm chain = %s' % b2m.get_id())\n",
    "\n",
    "    annotation.to_csv('../sandbox/detailed_mhc', index=None, sep='\\t')\n",
    "        \n",
    "#b2m_define(annotation)\n",
    "\n",
    "annotation = pd.read_csv('../sandbox/detailed_mhc', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "iclass = annotation[annotation.mhc_class == 'I']\n",
    "iclass = iclass[iclass.a_chain_allele.str.contains('[ABC]\\*', regex=True)]\n",
    "iclass.a_chain_allele = [':'.join(x.split(':')[:2]) for x in iclass.a_chain_allele]\n",
    "iclass = iclass.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "tmp = iclass[iclass.a_chain_allele.str.startswith('A*02:01')]\n",
    "with open('../sandbox/peptides_cluster', 'w') as file:\n",
    "    for i in range(tmp.shape[0]):\n",
    "        file.write('>%s\\n%s\\n' % (tmp.iloc[i, 0], tmp.iloc[i, 8]))\n",
    "        \n",
    "\n",
    "cluster1 = ['2x4r', '3gsv', '3gsq', '3mrd', '3mrb', '3mrc', '3gsx', '3gsw', '3gsu', '3gsr', '3mr9']\n",
    "cluster2 = ['4l3c', '1s9w', '3kla', '3gjf', '1s9y', '1s9x']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def align_pdb(table):\n",
    "    \n",
    "    class SelectChains(Bio.PDB.Select):\n",
    "        def __init__(self, chain_letters):\n",
    "            self.chain_letters = chain_letters\n",
    "        def accept_chain(self, chain):\n",
    "            return (chain.get_id() in self.chain_letters)\n",
    "\n",
    "    parser = Bio.PDB.PDBParser(QUIET=True)\n",
    "\n",
    "    name = table.iloc[0, 0]; amhc = table.iloc[0, 3]; pept = table.iloc[0, 7]; b2m = table.iloc[0, 5]\n",
    "    ref = parser.get_structure(id=name, file=pdb_path+'/'+name+'.pdb')[0]\n",
    "    ref_atoms = []\n",
    "    for atom in ref[amhc].get_atoms():\n",
    "        if atom.get_name() == 'CA':\n",
    "            ref_atoms.append(atom)\n",
    "\n",
    "    io = Bio.PDB.PDBIO()\n",
    "    io.set_structure(ref)\n",
    "    io.save('../sandbox/aligned_pdb/%s_cut.pdb' % name, SelectChains([amhc, pept, b2m]))\n",
    "\n",
    "    super_imposer = Bio.PDB.Superimposer()\n",
    "\n",
    "    for i in range(1, table.shape[0]):\n",
    "        name = table.iloc[i, 0]\n",
    "        amhc = table.iloc[i, 3]\n",
    "        b2m  = table.iloc[i, 5]\n",
    "        pept = table.iloc[i, 7]\n",
    "        sample = parser.get_structure(id=name, file=pdb_path+'/'+name+'.pdb')[0]\n",
    "        sample_atoms = []\n",
    "        for atom in sample[amhc].get_atoms():\n",
    "            if atom.get_name() == 'CA':\n",
    "                sample_atoms.append(atom)\n",
    "                \n",
    "        length = min(len(ref_atoms), len(sample_atoms))\n",
    "        super_imposer.set_atoms(ref_atoms[:length], sample_atoms[:length])\n",
    "        super_imposer.apply(sample.get_atoms())\n",
    "        \n",
    "        io.set_structure(sample)\n",
    "        io.save('../sandbox/aligned_pdb/%s_cut.pdb' % name, SelectChains([amhc, pept, b2m]))\n",
    "        \n",
    "#align_pdb(iclass.loc[:50, :])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pdb</th>\n",
       "      <th>mhc_class</th>\n",
       "      <th>host</th>\n",
       "      <th>a_chain_id</th>\n",
       "      <th>a_chain_allele</th>\n",
       "      <th>b_chain_id</th>\n",
       "      <th>b_chain_allele</th>\n",
       "      <th>antigen_id</th>\n",
       "      <th>antigen_seq</th>\n",
       "      <th>antigen_len</th>\n",
       "      <th>antigen_name</th>\n",
       "      <th>antigen_organism</th>\n",
       "      <th>comments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3i6l</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*24:02</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>QFKDNVILL</td>\n",
       "      <td>9</td>\n",
       "      <td>Nucleoprotein</td>\n",
       "      <td>SARS coronavirus Tor2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5swq</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>CVNGSCFTV</td>\n",
       "      <td>9</td>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>Influenza A virus</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>3mri</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>P</td>\n",
       "      <td>CINMWCWTV</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2uwe</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>H</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>I</td>\n",
       "      <td>B2M</td>\n",
       "      <td>J</td>\n",
       "      <td>ALWGFFPVL</td>\n",
       "      <td>9</td>\n",
       "      <td>chromosome 15 open reading frame 24</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>3gsw</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>P</td>\n",
       "      <td>NLVPMVAAV</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    pdb mhc_class   host a_chain_id a_chain_allele b_chain_id b_chain_allele  \\\n",
       "0  3i6l         I  Human          D        A*24:02          E            B2M   \n",
       "1  5swq         I  Human          A        A*02:01          B            B2M   \n",
       "5  3mri         I  Human          A        A*02:01          B            B2M   \n",
       "7  2uwe         I  Human          H        A*02:01          I            B2M   \n",
       "8  3gsw         I  Human          A        A*02:01          B            B2M   \n",
       "\n",
       "  antigen_id antigen_seq  antigen_len                         antigen_name  \\\n",
       "0          F   QFKDNVILL            9                        Nucleoprotein   \n",
       "1          C   CVNGSCFTV            9                        neuraminidase   \n",
       "5          P   CINMWCWTV            9                                  NaN   \n",
       "7          J   ALWGFFPVL            9  chromosome 15 open reading frame 24   \n",
       "8          P   NLVPMVAAV            9                                  NaN   \n",
       "\n",
       "        antigen_organism comments  \n",
       "0  SARS coronavirus Tor2      NaN  \n",
       "1      Influenza A virus      NaN  \n",
       "5                    NaN      NaN  \n",
       "7           Homo sapiens      NaN  \n",
       "8                    NaN      NaN  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iclass[(iclass.antigen_len == 9) & (iclass.a_chain_allele.str.startswith('A'))].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# if residue in mhc is close to antigen then return 1.\n",
    "# if further return 0.5\n",
    "# if far return 0.01\n",
    "def get_points(res, chain):\n",
    "    mdist = 1000\n",
    "    for a in res:\n",
    "        for b in chain.get_atoms():\n",
    "            dist = a-b\n",
    "            if dist < mdist:\n",
    "                mdist = dist\n",
    "                if mdist < 5:\n",
    "                    return 1.0\n",
    "    if mdist < 10:\n",
    "        return 0.5\n",
    "    return 0.01\n",
    "\n",
    "# return sorted list of best alleles\n",
    "def score_alignments(scores, blasted, points, alleles, line_id, size):\n",
    "    p1 = re.compile('\\w+\\s+([0-9]+)\\s+(\\w+)\\s+([0-9]+)\\\\n')\n",
    "    p2 = re.compile('Subject_([0-9]+)')\n",
    "    p3 = re.compile('[A-Z]')\n",
    "\n",
    "    res_start = start = end = counter = 0\n",
    "    \n",
    "    for line in blasted:\n",
    "        if line[:7] == 'Query_1':\n",
    "            m = p1.match(line)\n",
    "            res_start = int(m.group(1))-1\n",
    "            start = m.start(2)\n",
    "            end = m.end(2)\n",
    "            counter = 0\n",
    "            \n",
    "        if line[:8] == 'Subject_':\n",
    "            if counter >= size:\n",
    "                break\n",
    "            allele_id = int(p2.match(line).group(1))-1\n",
    "            chunk = line[start:end+1]\n",
    "            if np.isinf(scores[line_id, allele_id]):\n",
    "                scores[line_id, allele_id] = 0.0\n",
    "            \n",
    "            for m in p3.finditer(chunk):\n",
    "                res_ind = res_start+m.start()\n",
    "                scores[line_id, allele_id] += points[res_ind]\n",
    "            counter += 1\n",
    "\n",
    "    return scores\n",
    "\n",
    "# blast crystal hla I seqs against hla data base\n",
    "# and create allele-pdb compatibility table\n",
    "def make_allele2pdb_table():\n",
    "    p4 = re.compile('.+([ABC]\\*[0-9]+:[0-9A-Z]+)')\n",
    "    alleles = []\n",
    "    file = open(\"../data/mhc_alleles/hla/hla1class.fasta\", 'r')\n",
    "    lines = file.readlines()\n",
    "    file.close()\n",
    "    for line in lines:\n",
    "        if line[0] == '>':\n",
    "            #alleles.append(p4.search(line).group(1))\n",
    "            alleles.append(line.split(';')[1])\n",
    "\n",
    "    parser = Bio.PDB.PDBParser(QUIET=True)\n",
    "    ppb = Bio.PDB.CaPPBuilder()\n",
    "\n",
    "    size = len(alleles)\n",
    "    num_alignments = size\n",
    "\n",
    "    scores = np.zeros((iclass.shape[0], size))\n",
    "    scores.fill(np.inf)\n",
    "\n",
    "    for i in range(iclass.shape[0]):\n",
    "        name = iclass.iloc[i, 0]\n",
    "        print(name)\n",
    "        chain = iclass.iloc[i, 3]\n",
    "        antichain = iclass.iloc[i, 7]\n",
    "        struct = parser.get_structure(id=name, file=pdb_path+'/'+name+'.pdb')[0]\n",
    "        seq = ''.join([str(x.get_sequence()) for x in ppb.build_peptides(struct[chain])])\n",
    "        query = open('../sandbox/query.fasta', 'w')\n",
    "        query.write('>%s\\n%s\\n' % (name, seq))\n",
    "        query.close()\n",
    "\n",
    "        command = [\"blastp\", \"-query\", \"../sandbox/query.fasta\",\n",
    "                   \"-subject\", \"../data/mhc_alleles/hla/hla1class.fasta\", \"-out\", \n",
    "                   \"../sandbox/out\", \"-outfmt\", str(3), \"-num_alignments\", str(num_alignments),\n",
    "                   \"-line_length\", str(1000), \"-ungapped\", \"-comp_based_stats\", \"F\"]\n",
    "        if subprocess.call(command) != 0:\n",
    "            print('Failed to blast')\n",
    "            exit()\n",
    "\n",
    "        points = np.array([0.0]*len(struct[chain]))\n",
    "        residues = list(struct[chain])\n",
    "        for j in range(len(residues)):\n",
    "            points[j] = get_points(residues[j], struct[antichain])\n",
    "\n",
    "        file = open(\"../sandbox/out\", 'r')\n",
    "        blasted = file.readlines()\n",
    "        file.close()\n",
    "\n",
    "        scores = score_alignments(scores, blasted, points, alleles, i, num_alignments)\n",
    "\n",
    "    alleles_vs_pdb = pd.DataFrame([alleles]+[list(x) for x in list(scores)]).T.drop_duplicates(0)\n",
    "    alleles_vs_pdb.index = alleles_vs_pdb[0]\n",
    "    alleles_vs_pdb = alleles_vs_pdb.iloc[:, 1:]\n",
    "    alleles_vs_pdb.columns = list(iclass.pdb)\n",
    "    alleles_vs_pdb = alleles_vs_pdb.replace(np.inf, np.nan)\n",
    "    alleles_vs_pdb.to_csv('../sandbox/alleles_vs_pdb', sep='\\t', float_format='%.2f')\n",
    "    \n",
    "#make_allele2pdb_table()\n",
    "\n",
    "alleles_vs_pdb = pd.read_csv('../sandbox/alleles_vs_pdb', sep='\\t', index_col=0, low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ligand_assays = pd.read_csv('../data/epitopes/mhc_ligand_full.csv', skiprows=1, low_memory=False)\n",
    "ligand_assays = ligand_assays[ligand_assays['MHC allele class'] == 'I']\n",
    "\n",
    "affinity = ligand_assays[['Description', 'Allele Name', 'Qualitative Measure', 'Quantitative measurement']]\n",
    "affinity['Length'] = [len(x) for x in affinity.Description]\n",
    "affinity = affinity[(affinity.Length < 12) & (affinity.Length > 7)]\n",
    "affinity = affinity[affinity['Allele Name'].str.contains('HLA-[ABC]\\*\\S+$', regex=True)]\n",
    "affinity['Allele Name'] = [x[4:] for x in affinity['Allele Name']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# compare peptides with bias to terminal acids\n",
    "def cmp_peptides(a, b):\n",
    "        assert(len(a) == len(b))\n",
    "        score = 0.0\n",
    "        length = len(a)\n",
    "        for i in range(length):\n",
    "            if a[i] != b[i]:\n",
    "                if i in [0,1,length-1]:\n",
    "                    score += 1.0\n",
    "                else:\n",
    "                    score += 0.5\n",
    "        return score\n",
    "\n",
    "# create peptide-pdb compatibility table\n",
    "def make_peptide2pdb_table():\n",
    "    antigen_list = list(set(affinity['Description']))\n",
    "    size = len(antigen_list)\n",
    "\n",
    "    peptide_scores = np.zeros((size, iclass.shape[0]))\n",
    "    peptide_scores.fill(np.inf)\n",
    "\n",
    "    for counter in range(size):\n",
    "        if counter % 1000 == 0:\n",
    "            print(counter)\n",
    "        seq = antigen_list[counter]\n",
    "        length = len(seq)\n",
    "\n",
    "        local = iclass[iclass.antigen_len == length]\n",
    "        for i in local.index:\n",
    "            peptide_scores[counter, i] = 0.0\n",
    "            crystal_seq = local.loc[i, 'antigen_seq']\n",
    "            peptide_scores[counter, i] = cmp_peptides(crystal_seq, seq)\n",
    "\n",
    "    peptides_vs_pdb = pd.DataFrame(peptide_scores, index=antigen_list, columns=list(iclass.pdb))\n",
    "    peptides_vs_pdb = peptides_vs_pdb.replace(np.inf, np.nan)\n",
    "    peptides_vs_pdb.to_csv('../sandbox/peptides_vs_pdb', sep='\\t')\n",
    "\n",
    "#make_peptide2pdb_table()\n",
    "\n",
    "peptides_vs_pdb = pd.read_csv('../sandbox/peptides_vs_pdb', sep='\\t', index_col=0, low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# For each pdb in data calculate minimal rsmd of its peptide with a peptide of another pdb\n",
    "\n",
    "def peptides_rmsd(name1, chain1, name2, chain2):\n",
    "    parser = Bio.PDB.PDBParser(QUIET=True)\n",
    "    super_imposer = Bio.PDB.Superimposer()\n",
    "    \n",
    "    str1 = parser.get_structure(id=name1, file=pdb_path+'/'+name1+'.pdb')[0]\n",
    "    str1_atoms = []\n",
    "    for atom in str1[chain1].get_atoms():\n",
    "        if atom.get_name() == 'CA':\n",
    "            str1_atoms.append(atom)\n",
    "\n",
    "    str2 = parser.get_structure(id=name2, file=pdb_path+'/'+name2+'.pdb')[0]\n",
    "    str2_atoms = []\n",
    "    for atom in list(str2[chain2].get_atoms()):\n",
    "        if atom.get_name() == 'CA':\n",
    "            str2_atoms.append(atom)\n",
    "            \n",
    "    assert(len(str1_atoms) == len(str2_atoms))\n",
    "    length = len(str1_atoms)\n",
    "    super_imposer.set_atoms(str1_atoms, str2_atoms)\n",
    "    \n",
    "    return super_imposer.rms\n",
    "\n",
    "def peptide_min_rmsd_calc():\n",
    "    pdb2rmsd = {}\n",
    "    errors = []\n",
    "\n",
    "    for pdb in list(iclass.pdb):\n",
    "        print('\\n'+pdb+':')\n",
    "\n",
    "        index = iclass[iclass.pdb == pdb].index[0]\n",
    "        pepseq = iclass.loc[index, 'antigen_seq']\n",
    "        allele = iclass.loc[index, 'a_chain_allele']\n",
    "        chunk = alleles_vs_pdb.loc[alleles_vs_pdb.index.str.startswith(allele), :]\n",
    "        print(pepseq)\n",
    "        if pepseq not in peptides_vs_pdb.index:\n",
    "            continue\n",
    "\n",
    "        best_pdbs = (chunk.apply(min) + peptides_vs_pdb.loc[pepseq, :]/10).sort_values()\n",
    "\n",
    "        rmsd_list = []\n",
    "        for i in xrange(1, 10):\n",
    "            best_pdb = best_pdbs.index[i]\n",
    "            best_index = iclass[iclass.pdb == best_pdb].index[0]\n",
    "            print(best_pdb)\n",
    "            try:\n",
    "                rmsd = peptides_rmsd(pdb, iclass.loc[index, 'antigen_id'], best_pdb, iclass.loc[best_index, 'antigen_id'])\n",
    "            except AssertionError:\n",
    "                print('Assertion error: incorrect peptide length')\n",
    "                errors.append((pdb, best_pdb))\n",
    "\n",
    "            rmsd_list.append(rmsd)\n",
    "            #print(best_pdb+' gives '+str(rmsd))\n",
    "\n",
    "        pdb2rmsd.update({pdb: min(rmsd_list)})\n",
    "\n",
    "    pd.DataFrame(pdb2rmsd.values(), pdb2rmsd.keys()).to_csv('../sandbox/peptide_min_rsmd', header=None, sep='\\t')\n",
    "\n",
    "#peptide_min_rmsd_calc()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pdb</th>\n",
       "      <th>mhc_class</th>\n",
       "      <th>host</th>\n",
       "      <th>a_chain_id</th>\n",
       "      <th>a_chain_allele</th>\n",
       "      <th>b_chain_id</th>\n",
       "      <th>b_chain_allele</th>\n",
       "      <th>antigen_id</th>\n",
       "      <th>antigen_seq</th>\n",
       "      <th>antigen_len</th>\n",
       "      <th>antigen_name</th>\n",
       "      <th>antigen_organism</th>\n",
       "      <th>comments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3i6l</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>D</td>\n",
       "      <td>A*24:02:85</td>\n",
       "      <td>E</td>\n",
       "      <td>B2M</td>\n",
       "      <td>F</td>\n",
       "      <td>QFKDNVILL</td>\n",
       "      <td>9</td>\n",
       "      <td>Nucleoprotein</td>\n",
       "      <td>SARS coronavirus Tor2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5swq</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01:48</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>CVNGSCFTV</td>\n",
       "      <td>9</td>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>Influenza A virus</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3x14</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>B*08:02</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>FLRGRAYGL</td>\n",
       "      <td>9</td>\n",
       "      <td>nuclear antigen EBNA-3</td>\n",
       "      <td>Human herpesvirus 4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4jqx</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>B*44:03:33</td>\n",
       "      <td>C</td>\n",
       "      <td>B2M</td>\n",
       "      <td>B</td>\n",
       "      <td>EECDSELEIKRY</td>\n",
       "      <td>12</td>\n",
       "      <td>Trans-activator protein BZLF1</td>\n",
       "      <td>Human herpesvirus 4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5c0d</td>\n",
       "      <td>I</td>\n",
       "      <td>Human</td>\n",
       "      <td>A</td>\n",
       "      <td>A*02:01:48</td>\n",
       "      <td>B</td>\n",
       "      <td>B2M</td>\n",
       "      <td>C</td>\n",
       "      <td>AQWGPDPAAA</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    pdb mhc_class   host a_chain_id a_chain_allele b_chain_id b_chain_allele  \\\n",
       "0  3i6l         I  Human          D     A*24:02:85          E            B2M   \n",
       "1  5swq         I  Human          A     A*02:01:48          B            B2M   \n",
       "2  3x14         I  Human          A        B*08:02          B            B2M   \n",
       "3  4jqx         I  Human          A     B*44:03:33          C            B2M   \n",
       "4  5c0d         I  Human          A     A*02:01:48          B            B2M   \n",
       "\n",
       "  antigen_id   antigen_seq  antigen_len                   antigen_name  \\\n",
       "0          F     QFKDNVILL            9                  Nucleoprotein   \n",
       "1          C     CVNGSCFTV            9                  neuraminidase   \n",
       "2          C     FLRGRAYGL            9         nuclear antigen EBNA-3   \n",
       "3          B  EECDSELEIKRY           12  Trans-activator protein BZLF1   \n",
       "4          C    AQWGPDPAAA           10                            NaN   \n",
       "\n",
       "        antigen_organism comments  \n",
       "0  SARS coronavirus Tor2      NaN  \n",
       "1      Influenza A virus      NaN  \n",
       "2    Human herpesvirus 4      NaN  \n",
       "3    Human herpesvirus 4      NaN  \n",
       "4                    NaN      NaN  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "counter = 0\n",
    "for item in Bio.SeqIO.parse(\"../data/mhc_alleles/hla/hla1class.fasta\", 'fasta'):\n",
    "    if counter == 427:\n",
    "        print(item)\n",
    "        break\n",
    "    counter+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "mutate('4i48', alleledict['A*02:01'].seq, 'NLVPMVHTV', iclass)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'3mrb', '4i48'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "iclass[iclass.pdb == '3mrb']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "table=iclass\n",
    "pdb ='4hx1'\n",
    "parser = Bio.PDB.PDBParser(QUIET=True)\n",
    "ppb=Bio.PDB.PPBuilder()\n",
    "name = pdb; amhc = table[table.pdb == pdb].iloc[0, 3]; pept = table[table.pdb == pdb].iloc[0, 7]\n",
    "struct = parser.get_structure(id=name, file=pdb_path+'/'+name+'.pdb')[0]\n",
    "oldseq = str(ppb.build_peptides(struct[amhc])[0].get_sequence())\n",
    "oldpep = table[table.pdb == pdb].iloc[0, 8]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(' ', 1, ' ')"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res = ppb.build_peptides(struct[amhc])[0][0]\n",
    "res.get_id()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "str(ppb.build_peptides(struct[amhc])[0].get_sequence())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "affinity.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "alleledict.keys()[427]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "command = [\"blastp\", \"-query\", \"../testq\",\n",
    "                   \"-subject\", \"../tests\", \"-out\", \n",
    "                   \"../testout\", \"-outfmt\", str(3), \"-num_alignments\", str(20),\n",
    "                   \"-line_length\", str(1000), \"-ungapped\", \"-comp_based_stats\", \"F\"]\n",
    "if subprocess.call(command) != 0:\n",
    "    print('Failed to blast')\n",
    "    exit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda env:python2]",
   "language": "python",
   "name": "conda-env-python2-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
